{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1cWdsyOsrrZnrXUul6lCw1q88jay7SlsD","timestamp":1683731324320}],"toc_visible":true,"authorship_tag":"ABX9TyORh62TMFapQIDhBMV0i6Bb"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Day 1 - Introduction to Python Programming for Biologists\n","All commands for today's session are found below, except where you see the word ***EXAMPLE*** which may be just non functional psuedo-code to explain a concept.\n","\n","If you see ***Exercises*** then there is a challenge for you to write some code yourself!\n","\n","## Day 1\n","- Introduction to Python concepts and coding in notebooks\n","- Basic data types and simple functions\n","- Data manipulation and I/O handling      \n","- **More data types!<-------------**\n","- Project: Creating a program with variable inputs to manipulate DNA sequences"],"metadata":{"id":"QfYvI-_yE10I"}},{"cell_type":"markdown","source":["# More Data formats\n","\n","So far we've just looked at singe pieces of data whether they are individual numbers, or large strings of text. However the real power comes from collections of data in lists, tuples, and dictionaries.\n","\n"],"metadata":{"id":"chQY6saIQI0B"}},{"cell_type":"markdown","source":["# Lists\n","\n","A list is a collection of items that are ordered and changeable. Lists are one of the most commonly used data structures in Python and are used to store a collection of related values.\n","\n","To create a list in Python, we simply use square brackets and separate the items in the list with commas:"],"metadata":{"id":"O3dpWcuaRuhi"}},{"cell_type":"code","source":["# Create a list of DNA bases\n","bases = ['A', 'T', 'G', 'C', 'G', 'C']\n","\n","print(bases)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ehgMJYEoSVFS","executionInfo":{"status":"ok","timestamp":1683719171599,"user_tz":-60,"elapsed":4,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"a5053e75-e622-4e0b-bb5d-5a3015b963be"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['A', 'T', 'G', 'C', 'G', 'C']\n"]}]},{"cell_type":"markdown","source":["We can access individual elements in a list by using their index. In Python, list indexes start at 0, so the first element in the list has an index of 0, the second element has an index of 1, and so on. \n","\n","When selecting multiple elements the index must be inclusive of the last element to be included."],"metadata":{"id":"uE0oQ8hHSZXh"}},{"cell_type":"code","source":["# Access the first element in the list\n","print(bases[0])\n","\n","# Access the second, third, and fourth element in the list\n","print(bases[1:4])\n","\n","# Access from the third to the end of the list\n","print(bases[2:])\n","\n","# Access the last element in the list\n","print(bases[-1])\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kJbgZbocSaCh","executionInfo":{"status":"ok","timestamp":1683719176491,"user_tz":-60,"elapsed":189,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"57b75c0e-6632-4ff2-81a6-5ff5c4ff7b52"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["A\n","['T', 'G', 'C']\n","['G', 'C', 'G', 'C']\n","C\n"]}]},{"cell_type":"markdown","source":["Once you have data in your list, you can replace individual elements, or add, remove, insert etc."],"metadata":{"id":"DIuyHBqUSw1r"}},{"cell_type":"code","source":["# Modify the third element in the list\n","bases[2] = 'N'\n","\n","print(bases)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5PSpVqG6TBOD","executionInfo":{"status":"ok","timestamp":1683719210583,"user_tz":-60,"elapsed":188,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"3ec891d1-34e7-434d-b7ad-d2b4a6489047"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['A', 'T', 'N', 'N', 'C', 'G', 'C', 'U']\n"]}]},{"cell_type":"code","source":["# Add an element to the end of the list\n","bases.append('U')\n","\n","# Insert an element at a specific index\n","bases.insert(4, 'N')\n","\n","print(bases)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Yf8dU5JSS7LS","executionInfo":{"status":"ok","timestamp":1683719212630,"user_tz":-60,"elapsed":182,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"25f28646-7fd0-41d9-84ca-a9822d632d0b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['A', 'T', 'N', 'N', 'N', 'C', 'G', 'C', 'U', 'U']\n"]}]},{"cell_type":"code","source":["# Remove an element from the list - only the first instance!\n","bases.remove('N')\n","\n","# Remove the last element from the list\n","bases.pop()\n","\n","print(bases)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"z2eb7ewqThj7","executionInfo":{"status":"ok","timestamp":1683719219877,"user_tz":-60,"elapsed":218,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"2c3cfeed-9541-4578-8c39-9f75f1901eab"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['A', 'T', 'C', 'G']\n"]}]},{"cell_type":"markdown","source":["You can also add mulptiple elements on at the same time, or combine lists easily:"],"metadata":{"id":"_noxi2Life1X"}},{"cell_type":"code","source":["confirmed_genes = ['BRCA1', 'TP53', 'EGFR', 'KRAS']\n","suggested_target_genes = ['MET1', 'ROS']\n","\n","all_genes = confirmed_genes + suggested_target_genes\n","\n","print(all_genes)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"bNS7-RlQflWW","executionInfo":{"status":"ok","timestamp":1683714763366,"user_tz":-60,"elapsed":192,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"43e03efb-390c-4bdd-b816-85a6ab7cc8f4"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['BRCA1', 'TP53', 'EGFR', 'KRAS', 'MET1', 'ROS']\n"]}]},{"cell_type":"markdown","source":["## Exercise - BugFix\n","\n","Try to find and correct the issue in the following code blocks:"],"metadata":{"id":"-KlsVZIqEJT5"}},{"cell_type":"code","source":["# Our gene as a list of bases\n","my_insert = ['C', 'A', 'T', 'G', 'C', 'G', 'T', 'A', 'T', 'A', 'T', 'G', 'C', 'C', 'A', 'T', 'C', 'G']\n","\n","# Get last base\n","length = len(my_insert)\n","last_base = my_inser[length]\n","\n","print(last_base)\n","\n","# Remove all N's\n","my_insert.remove(\"N\")\n","\n","print(my_insert)"],"metadata":{"id":"_pq0YNuNEIjQ","colab":{"base_uri":"https://localhost:8080/","height":235},"executionInfo":{"status":"error","timestamp":1684542273569,"user_tz":-60,"elapsed":259,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"d84dc7d3-c6ac-4b1b-856d-a17157ad2a96"},"execution_count":2,"outputs":[{"output_type":"error","ename":"IndexError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m<ipython-input-2-7d8bed9d1586>\u001b[0m in \u001b[0;36m<cell line: 6>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# Get last base\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mlength\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmy_insert\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mlast_base\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmy_insert\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlength\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlast_base\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mIndexError\u001b[0m: list index out of range"]}]},{"cell_type":"markdown","source":["## Exercise - Manipulating lists"],"metadata":{"id":"IyZbrc6pOB0g"}},{"cell_type":"markdown","source":["We know that the restriction enzyme **DANase** always cuts 9bp after the ATG start codon. We want to create two list variables for the begginging and remaining part of the sequence and print them out.\n","\n","Fix and complete this code. \n","(Hint: In this example it should cut at CAT|GCA)"],"metadata":{"id":"79lEj84kGhPC"}},{"cell_type":"code","source":["# Our gene as a list of bases\n","my_gene = ['A', 'T', 'G', 'C', 'C', 'G', 'C', 'G', 'G', 'C', 'A', 'T', 'G', 'C', 'A', 'T', 'G', 'C', 'G', 'T', 'A', 'T', 'A', 'T', 'G', 'C', 'C', 'A', 'T', 'C', 'G']\n","\n","# Expected cut site: CAT|GCA\n","first_part = my_gene[]\n","second_part = my_gene[]\n","\n","print(first_part)\n","print(second_part)"],"metadata":{"id":"CWFQfEMEGA66"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can use the .join() method to change a list into a string for outputting\n","Make an easily readable output highlighing the restriction cut in our sequence so that it looks like:\n","```...ATCG-CGAT...```\n"],"metadata":{"id":"hgm8BqzDKZJ7"}},{"cell_type":"code","source":["# To start you off\n","print(' -space- '.join(first_part))"],"metadata":{"id":"sQHED2-PKk9S"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["---\n","\n","# Tuples\n","A tuple is a collection of ordered and immutable elements. Tuples are similar to lists, but **once a tuple is created, its elements cannot be modified**. Because they won't be sorted, elonogated or shortened, Tuples are often used to store related values that are constant and do not change. A classic example would be genomic co-ordinates.\n","\n","Because they are immutable it can be useful for storing data that should not be changed once read in, and are more memory-efficient and faster to access and process than lists.\n","\n","Tuples are differentiated from lists by using round brackets (parentheses) instead. Because we define a specific format we can access elements like so:"],"metadata":{"id":"RGWx6mFXUBQ6"}},{"cell_type":"code","source":["# Define a genomic coordinate as a tuple\n","TFBS_coords = ('TFBS01', 1500, 1700)\n","\n","# Print the chromosome name\n","print(TFBS_coords[0])\n","\n","# Print the start position\n","print(TFBS_coords[1])\n","\n","# Print the end position\n","print(TFBS_coords[2])\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"32i4KcYAUwb5","executionInfo":{"status":"ok","timestamp":1683732448646,"user_tz":-60,"elapsed":4,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"20dfccd4-01f3-4ffe-bef2-d08c8a7d8e81"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["TFBS01\n","1500\n","1700\n"]}]},{"cell_type":"markdown","source":["Because of the fixed order of tuples there is also a very useful function to quickly assign the values to variables:"],"metadata":{"id":"TyArylsGQ919"}},{"cell_type":"code","source":["id, start, end = TFBS_coords\n","\n","print(\"ID:\\t\", id)\n","print(\"Start:\\t\", start)\n","print(\"End:\\t\", end)\n","print(\"\\nBED format output:\", id + \":\" + str(start) + \"-\" + str(end))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"qakt6KEOQ9gu","executionInfo":{"status":"ok","timestamp":1683732630933,"user_tz":-60,"elapsed":275,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"beed1bc3-f462-4a62-d857-b396f026fe8c"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["ID:\t TFBS01\n","Start:\t 1500\n","End:\t 1700\n","\n","BED format output: TFBS01:1500-1700\n"]}]},{"cell_type":"markdown","source":["---\n","\n","# Ranges\n","Really, ranges are a type of automatic list method, where you can define the range and period in your sequence. They are not really a different data type, but very useful!\n","\n","You can easily create a simple range like so, but note that it is stored as a range-object until you force it to be a list:"],"metadata":{"id":"pVHT436sXvYe"}},{"cell_type":"code","source":["x = range(20)\n","\n","# As it's own object\n","print(x)\n","\n","# Forced List\n","print(list(x))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yEDP7O52Y4jW","executionInfo":{"status":"ok","timestamp":1683717654010,"user_tz":-60,"elapsed":871,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"cca647c4-3c2e-4356-ed54-038e6cb69e20"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["range(0, 20)\n","[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]\n"]}]},{"cell_type":"markdown","source":["You can also define the interval that you want your sequence to have"],"metadata":{"id":"sTzljupXc-09"}},{"cell_type":"code","source":["my_timeline = range(0, 20, 3)\n","print(list(my_timeline))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0M6iGIbCc6zW","executionInfo":{"status":"ok","timestamp":1683718709545,"user_tz":-60,"elapsed":186,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"62e32b86-5231-4c40-ba69-6be490b9f1c9"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[0, 3, 6, 9, 12, 15, 18]\n"]}]},{"cell_type":"markdown","source":["## Exercise - Creating ranges\n","\n","You have taken a flourescence measurement of your culturing algae every **6 minutes for 3 hours** to see the response to a growth stimulation. \n","\n","1. How many elements are going to be in your list? (You could either calculate it, or count the length of your measurements)\n","2. Create a list object named ```timeline```\n","3. Run the graph making code to plot your data\n","\n","Note: Remember that list indexes must be inclusive!"],"metadata":{"id":"73tE45DaZP-w"}},{"cell_type":"code","source":["# Your data\n","measurements = [0.68, 1.69, 3.37, 5.61, 7.89, 9.86, 10.9, 10.84, 9.86, 8.22, 6.46, 4.69, 3.19, 2.05, 1.24, 0.7, 0.38, 0.2, 0.1, 0.05, 0.02, 0.01, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 1.0, 2.0, 3.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]\n","\n","## Create your timeline\n","timeline = \n"],"metadata":{"id":"ni-XFz3IZrJ4","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1683733202040,"user_tz":-60,"elapsed":11,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"d1063111-d6b7-4bd3-da26-d291600830ce"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["51"]},"metadata":{},"execution_count":10}]},{"cell_type":"code","source":["# Make a quick graph\n","import matplotlib.pyplot as plt\n","\n","def plot_data(x, y):\n","    plt.plot(x, y)\n","    plt.xlabel('Time')\n","    plt.ylabel('Fluorescence')\n","    plt.show()\n","\n","plot_data(timeline, measurements)"],"metadata":{"id":"kdmwllxUUGJ3"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["---\n","\n","# Two Dimensional (2D) lists \n","\n","2d lists are where each element of the list is a list itself, representing a row of the annotation table. \n","\n","Here the first element in each row is the gene ID, the second is a confidence value, and the third is the length of the gene. \n","\n","We can access individual elements of the annotation table using two indices by combining square bracket indexing like with a normal list"],"metadata":{"id":"dHQW6joOSzSV"}},{"cell_type":"code","source":["gene_confs = [[0.92, 'MET1'], [0.82, 'EGFR'], [0.93, 'KRAS'], [0.4, 'TP53'], [0.94, 'ROS5'], [0.87, 'BRCA1']]\n","\n","print(gene_confs[1][1])\n","print(gene_confs[3][0])\n","print(gene_confs[5][-1])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"y80yavPgVvQm","executionInfo":{"status":"ok","timestamp":1683734628703,"user_tz":-60,"elapsed":4,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"3b641423-2aa6-4288-d12c-369dc114faa0"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[[0.92, 'MET1'], [0.82, 'EGFR'], [0.93, 'KRAS'], [0.4, 'TP53'], [0.94, 'ROS5'], [0.87, 'BRCA1']]\n"]}]},{"cell_type":"markdown","source":["We can then then add, remove, or sort the list as if it was a single list."],"metadata":{"id":"IN0PIuhlZvKG"}},{"cell_type":"markdown","source":["There is also a method named ```zip``` which can combine lists into an array. We will come back to this later in the course but works like this (note the requirement to force a ```list``` object for printing)\n"],"metadata":{"id":"tWtcaTq6aOzg"}},{"cell_type":"code","source":["bases = [\"A\", \"T\", \"C\", \"G\"]\n","count = [24, 17, 73, 26]\n","\n","combined_base_count = zip(bases, count)\n","\n","print(combined_base_count)\n","print(list(combined_base_count))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"epyWZLAGasNA","executionInfo":{"status":"ok","timestamp":1683734934636,"user_tz":-60,"elapsed":8,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"bab36286-da9a-428b-a627-c3531ae78963"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["<zip object at 0x7f72335a5600>\n","[('A', 24), ('T', 17), ('C', 73), ('G', 26)]\n"]}]},{"cell_type":"markdown","source":["It can also be sometimes useful to do this in reverse, splitting a 2D list into two lists:"],"metadata":{"id":"k4B5OynX93p9"}},{"cell_type":"code","source":["line_exp = [(\"WT\", 98), (\"Mut1\", 76), (\"Mut2\", 86), (\"Mut3\", 79)]\n","\n","line, exp = zip(*line_exp)\n","\n","print(line)\n","print(exp)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"TWeRorxK9-Jd","executionInfo":{"status":"ok","timestamp":1683979100002,"user_tz":-60,"elapsed":6,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"2692be23-933f-4bda-b36c-0eb86ce97f69"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["('WT', 'Mut1', 'Mut2', 'Mut3')\n","(98, 76, 86, 79)\n"]}]},{"cell_type":"markdown","source":["## Exercise: Manipulating 2D lists\n","\n","1. Sort the gene list confidence value - sort() will only work on the first element in the sublists\n","2. Create a new variable named ```highest``` with the gene ID with the highest confidence value and print it\n","3. We have new data! Add the gene ```APC1``` with the conf value ```0.97``` and the gene ```COI``` with ```0.72``` to the list\n","4. Lets put the three genes with the lowest confidence into a new list named ```low_conf_genes```\n","5. Actually, even better lets add a new element to the low confidence genes saying ```RETEST```\n","6. Print out just the list of conf_values\n","7. Calculate the average of the values. You can use the ```sum()``` and ```len()``` functions on a list like this.\n","\n","Note: If you have done programming before you may notice that in several of these examples there are better and quicker ways to achieve the steps by using loops and conditionals. We will cover this in the next session but for now, use the commands we've learned\n"],"metadata":{"id":"21ou3cQMXQj-"}},{"cell_type":"code","source":["gene_confs = [[0.92, 'MET1'], [0.62, 'EGFR'], [0.93, 'KRAS'], [0.74, 'TP53'], [0.94, 'ROS5'], [0.87, 'BRCA1']]\n"],"metadata":{"id":"PF0FvjEqeaZ2","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1683736018127,"user_tz":-60,"elapsed":3,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"0ff6f999-39db-4fc2-cd99-bb7132682a96"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Highest confidence gene is: ROS5\n","[[0.62, 'EGFR'], [0.74, 'TP53'], [0.78, 'COI'], [0.87, 'BRCA1'], [0.92, 'MET1'], [0.93, 'KRAS'], [0.94, 'ROS5'], [0.97, 'APC1']]\n","[[0.62, 'EGFR', 'RETEST'], [0.74, 'TP53', 'RETEST'], [0.78, 'COI', 'RETEST'], [0.87, 'BRCA1'], [0.92, 'MET1'], [0.93, 'KRAS'], [0.94, 'ROS5'], [0.97, 'APC1']]\n"]}]},{"cell_type":"markdown","source":["---\n","\n","# Dictionaries\n","\n","A dictionary at first looks like a 2D list but is actually unique because all data is in unordered key-value pairs. They are particularly useful (like a real dictionary!) for finding the corresponding pair when you have one value for example:\n","\n","- ```gene => sequence```\n","- ```aminoAcid => frequency```\n","\n","This dictionary has a frequency count of each base, where the first element in each pair is the \"key\" reference, and the second element is the \"value\". \n","\n","It's defined in the format of 'key' : 'value', between braces (curly brackets)"],"metadata":{"id":"rQ5nBszYVmZT"}},{"cell_type":"code","source":["# Create a dictionary of DNA base counts\n","base_counts = {'A': 101, 'T': 250, 'G': 125, 'C': 92}\n","print(base_counts)\n","\n","# Print just the values\n","print(base_counts.values())\n","\n","# Print just the keys\n","print(base_counts.keys())"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"rTmf-Cp8WHSi","executionInfo":{"status":"ok","timestamp":1683737972712,"user_tz":-60,"elapsed":406,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"630a8623-c2ff-4af5-e8e3-55a048adfde0"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["{'A': 101, 'T': 250, 'G': 125, 'C': 92}\n","dict_values([101, 250, 125, 92])\n","dict_keys(['A', 'T', 'G', 'C'])\n"]}]},{"cell_type":"markdown","source":["Note how the ```.keys()``` and ```.values()``` methods return dict objects (just like when we made a range). They can be turned into a list using the ```list()``` function in the same way.\n","\n","I'll point out here, we can also use the ```.items()``` method to return key:value pairs as a tuple, but that is for more complicated functions we will look at later.\n","\n","We can then access individual values based on their key, or modify/delete them:"],"metadata":{"id":"cw8OZ_CwWvxT"}},{"cell_type":"code","source":["# Print corresponding value\n","print(\"Number of Adenine:\", base_counts['A'])\n","print(\"Number of Thymine:\", base_counts['T'])\n","\n","# Modify the value for a key\n","base_counts['A'] = 65\n","\n","# Add a new key-value pair to the dictionary\n","base_counts['N'] = 5\n","\n","# Remove the 'C' key-value pair from the dictionary\n","del base_counts['C']\n","\n","print(base_counts)\n"],"metadata":{"id":"JMKJ30HtW07q"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Dictionaries are really important and powerful. Lets look at another example where we can attach specific information to geneIDs, and use the ```.get``` method to search the dictionary.\n","\n","Here we can combine a gene sequence dictionary with our 2D list exercise from earlier"],"metadata":{"id":"KsVBxlMZXqI7"}},{"cell_type":"code","source":["# Dictionary of gene names and sequences\n","gene_dict = {'BRCA1': 'ATGTTGTCATCGTTGAGCTTTGCTTCCT',\n","             'TP53': 'ATGGAGGAGCCGCAGTCAGATC',\n","             'EGFR': 'ATGACCATCCAAGATGATGGTGTC',\n","             'KRAS': 'ATGACTGAATATAAACTTGTGGTAG',\n","             'BRAF': 'ATGGTCCAGCTTGGACCCACTCC',\n","             'ALK': 'ATGAAGGAGCCCTCAGATTTCTTG',\n","             'RET': 'ATGGGTGGGTTGTCGGAAGATCTT',\n","             'ROS1': 'ATGAGCCACCCAGGTCCCTGTAGT',\n","             'MET': 'ATGGCTTCAAGCTGTTGTCGTGAAGA'}\n","\n","# gene confidence values\n","gene_confs = [[0.92, 'MET1'], [0.62, 'EGFR'], [0.93, 'KRAS'], [0.74, 'TP53'], [0.94, 'ROS5'], [0.87, 'BRCA1']]\n","\n","# sort and get lowest conf gene ID\n","sorted_gene_confs = sorted(gene_confs)\n","lowest_gene_ID = sorted_gene_confs[0][1]\n","\n","print(lowest_gene_ID)\n","\n","# Search dictionary keys for that ID\n","print(gene_dict.get(lowest_gene_ID))\n","\n","# easier to read!\n","print(\"Gene sequence for\", lowest[1], \"(Confidence value:\", str(lowest[0]) + \") is:\",  gene_dict.get(lowest[1]))\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"O8UFUTAvXucr","executionInfo":{"status":"ok","timestamp":1684185982774,"user_tz":-60,"elapsed":461,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"248899c8-e69c-43ff-8bfe-bfed4f2c1966"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["EGFR\n","ATGACCATCCAAGATGATGGTGTC\n","Gene sequence for EGFR (Confidence value: 0.62) is: ATGACCATCCAAGATGATGGTGTC\n"]}]},{"cell_type":"markdown","source":["Actually, lets throw away the lowest value. We can use ```.pop()``` in the same way as with a list. Lets throw away the bottom three:"],"metadata":{"id":"D3u6X12NSE_s"}},{"cell_type":"code","source":["print(sorted_gene_confs)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"zZgqTCKuSeLv","executionInfo":{"status":"ok","timestamp":1684186016531,"user_tz":-60,"elapsed":225,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"55b2c2ec-6d1b-4dc6-b172-b7b39e8baa67"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[[0.62, 'EGFR'], [0.74, 'TP53'], [0.87, 'BRCA1'], [0.92, 'MET1'], [0.93, 'KRAS'], [0.94, 'ROS5']]\n"]}]},{"cell_type":"code","source":["print(\"gene_dict contains\", len(gene_dict), \"genes\")\n","gene_dict.pop(\"EGFR\")\n","gene_dict.pop(\"TP53\")\n","gene_dict.pop(\"BRCA1\")\n","print(\"gene_dict contains\", len(gene_dict), \"genes\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jA5MCmWZTGIp","executionInfo":{"status":"ok","timestamp":1684185985502,"user_tz":-60,"elapsed":211,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"26078b22-0c96-4675-bc3e-7a305a939f0b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["gene_dict contains 9 genes\n","gene_dict contains 6 genes\n"]}]},{"cell_type":"markdown","source":["Actually, even better idea! Lets put the lowest three into a new list called ```bad_genes```. \n","\n","Note, what happens if you run this code immediately? Read the error and identify what is wrong, and correct it."],"metadata":{"id":"2-4AuHdaT1Xk"}},{"cell_type":"code","source":["bad_genes = []\n","\n","print(\"gene_dict contains\", len(gene_dict), \"genes\")\n","bad_genes += gene_dict.pop(\"EGFR\")\n","bad_genes += gene_dict.pop(\"TP53\")\n","bad_genes += gene_dict.pop(\"BRCA1\")\n","print(\"gene_dict contains\", len(gene_dict), \"genes\")"],"metadata":{"id":"yfGTKzuPT8Hy"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["There are lots of powerful methods to search throughnot just the keys, but also the values to find relevant keys which can be very useful but first we need to learn a bit about loops and conditionals!"],"metadata":{"id":"QKAudoMCYk2t"}},{"cell_type":"markdown","source":["## Exercises\n","\n","1. Create a dictionary of bacteria and confluence values using the data:\n","\n","```\n","[\"E. coli\", \"S. aureus\", \"P. aeruginosa\", \"K. pneumoniae\", \"A. baumannii\"]\n","\n","[60, 80, 75, 90, 70]\n","```\n","2. Print the confluence value of *K. pneumoniae* from the dictionary\n","3. A function we haven't used yet is ```sum()```, but it works just like ```len()```. Use both of these to calculate the average confluence of all the samples in the dictionary\n","\n","\n"],"metadata":{"id":"xKbU1nRZkgP5"}},{"cell_type":"code","source":["# Format reminder\n","bacteria_dict = {\"E. coli\": 60, ........}"],"metadata":{"id":"8JOxai9ZlReC"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YWM78byVlInR","executionInfo":{"status":"ok","timestamp":1683738235852,"user_tz":-60,"elapsed":13,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"a20f086a-8d29-4b55-e53a-e25c19d8d78d"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["{'E. coli': 60, 'S. aureus': 80, 'P. aeruginosa': 75, 'K. pneumoniae': 90, 'A. baumannii': 70}\n","90\n","75.0\n"]}]}]}