{"cells":[{"cell_type":"markdown","metadata":{"id":"QfYvI-_yE10I"},"source":["# Practical Python Programming for Biologists\n","Author: Dr. Daniel Pass | www.CompassBioinformatics.com\n","\n","---"]},{"cell_type":"markdown","metadata":{"id":"chQY6saIQI0B"},"source":["# More Data formats\n","\n","So far we've just looked at singe pieces of data (as individual numbers, strings of text, or True/False values). However the real power comes from collections of data in lists, tuples, and dictionaries."]},{"cell_type":"markdown","metadata":{"id":"O3dpWcuaRuhi"},"source":["# Lists\n","\n","A list is a collection of items that are ordered and changeable. Lists are one of the most commonly used data structures in Python and are used to store a collection of related values.\n","\n","To create a list in Python, we simply use square brackets and separate the items in the list with commas:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":224,"status":"ok","timestamp":1684696371235,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"ehgMJYEoSVFS","outputId":"274cbfd7-4e5f-4dad-d3c8-b724a93f684c"},"outputs":[],"source":["# Create a list of DNA bases\n","bases = ['A', 'T', 'G', 'C', 'G', 'C']\n","\n","print(bases)"]},{"cell_type":"markdown","metadata":{"id":"uE0oQ8hHSZXh"},"source":["Because they are stored in order we can access individual elements in a list by using their index. Indexes always start at 0, so the first element in the list has an index of 0, the second element has an index of 1, and so on. \n","\n","When selecting multiple elements the index must be inclusive of the last element to be included. Note that it will return a list too, not just the elements you ask for."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":233,"status":"ok","timestamp":1684696384211,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"kJbgZbocSaCh","outputId":"1d663f26-3a00-42d7-8f5e-2046558a7a92"},"outputs":[],"source":["# Access the first element in the list\n","print(bases[0])"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Access the second, third, and fourth element in the list\n","print(bases[1:4])"]},{"cell_type":"markdown","metadata":{},"source":["There are also some useful shortuts, for example if you leave one value empty it will assume start/end of the list. You can also use the minus ```-``` character to start from the end of the list."]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["\n","# Access from the third to the end of the list\n","print(bases[2:])\n","\n","# Access the last element in the list\n","print(bases[-1])\n"]},{"cell_type":"markdown","metadata":{"id":"DIuyHBqUSw1r"},"source":["Once you have data in your list, you can easily replace individual elements, or add, remove, insert etc."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":299,"status":"ok","timestamp":1684696385429,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"5PSpVqG6TBOD","outputId":"dffc32fb-1500-4546-94f7-60729fcb3ef3"},"outputs":[],"source":["# Modify the third element in the list\n","bases[2] = 'N'\n","\n","print(bases)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":281,"status":"ok","timestamp":1684696386518,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"Yf8dU5JSS7LS","outputId":"c746a613-dd50-4d41-8ce1-6b8876ba0a6d"},"outputs":[],"source":["# Add an element to the end of the list\n","bases.append('U')\n","\n","# Insert an element at a specific index\n","bases.insert(4, 'N')\n","\n","print(bases)\n","\n","# Remove an element from the list - only the first instance!\n","bases.remove('N')\n","print(bases)"]},{"cell_type":"markdown","metadata":{"id":"E4fjTbctuGBg"},"source":["The ```.pop``` method is useful as it both takes the last element (by default) and removes it, but also can put it into a new variable."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":341,"status":"ok","timestamp":1684696387834,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"z2eb7ewqThj7","outputId":"496de102-747d-4e75-d340-1162702feb52"},"outputs":[],"source":["# A list of sites we still have to process\n","sites = [\"CDF\", \"BRI\", \"LDN\", \"SWN\", \"EXT\"]\n","\n","# Remove the last element from the list (and discard it)\n","print(sites)\n","sites.pop()\n","print(sites)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# A list of sites we still have to process\n","sites = [\"CDF\", \"BRI\", \"LDN\", \"SWN\", \"EXT\"]\n","\n","last_site = sites.pop()\n","print(\"The most recent site was\", last_site)\n","print(\"Remaining sites to be processed are:\", sites)"]},{"cell_type":"markdown","metadata":{"id":"_noxi2Life1X"},"source":["You can also add mulptiple elements on at the same time, or combine lists easily:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":212,"status":"ok","timestamp":1684696396079,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"bNS7-RlQflWW","outputId":"06420db5-9371-4a72-e730-83ad7b1619b3"},"outputs":[],"source":["confirmed_genes = ['BRCA1', 'TP53', 'EGFR', 'KRAS']\n","suggested_target_genes = ['MET1', 'ROS']\n","\n","all_genes = confirmed_genes + suggested_target_genes + ['ABC13']\n","\n","print(all_genes)"]},{"cell_type":"markdown","metadata":{"id":"-KlsVZIqEJT5"},"source":["## Exercise - BugFix\n","\n","Try to find and correct the issues in the code block:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":235},"executionInfo":{"elapsed":259,"status":"error","timestamp":1684696745125,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"_pq0YNuNEIjQ","outputId":"919d7032-07a7-4825-9a6e-940d22d88989"},"outputs":[],"source":["# Our gene insert as a list of bases\n","my_insert = ['C', 'A', 'T', 'G', 'C', 'G', 'T', 'A', 'T', 'A', 'T', 'N', 'N', 'C', 'A', 'T', 'C', 'G']\n","\n","# Get last base\n","region_length = len(my_insert)\n","last_base = my_insert[length]\n","\n","print(last_base)\n","\n","# Remove all N's\n","my_insert.remove(\"N\")\n","\n","print(my_insert)"]},{"cell_type":"markdown","metadata":{"id":"79lEj84kGhPC"},"source":["## Exercise - Manipulating lists\n","\n","We know that the restriction enzyme **DANase** always cuts 9bp after the start codon. We want to create two list variables for the begginging and remaining part of the sequence and print them out.\n","\n","1. Complete this code. Using numerical slicing, cut the string 9 bases ***after*** ATG"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"CWFQfEMEGA66"},"outputs":[],"source":["# Our gene as a list of bases\n","my_gene = ['A', 'T', 'G', 'C', 'C', 'G', 'C', 'G', 'G', 'C', 'A', 'T', 'G', 'G', 'A', 'C', 'G', 'C', 'G', 'T', 'A', 'T', 'A', 'T', 'G', 'C', 'C', 'A', 'T', 'C', 'G']\n","\n","# Expected cut site: CAT|GCA\n","first_part = my_gene[]\n","second_part = my_gene[]\n","\n","print(first_part)\n","print(second_part)"]},{"cell_type":"markdown","metadata":{},"source":["<details>\n","<summary>Expected output</summary>\n","\n","In this example it should cut at (```CAT|GGA```)\n","\n","</details>"]},{"cell_type":"markdown","metadata":{"id":"hgm8BqzDKZJ7"},"source":["Extension: We can use the .join() method to join all the elements in a list into a string and to make it easier for reading. The part between the single quotes is what will be between each element.\n","\n","2. Make an easily readable output highlighing the restriction cut in our sequence so that it looks like:\n","```...ATCG-CGAT...```\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"sQHED2-PKk9S"},"outputs":[],"source":["# To start you off\n","print(' - '.join(first_part))"]},{"cell_type":"markdown","metadata":{"id":"RGWx6mFXUBQ6"},"source":["---\n","\n","# Tuples\n","A tuple is a collection of \"ordered and immutable\" elements. Basically they are similar to lists, but **once a tuple is created, its elements cannot be changed**. \n","\n","Because they won't be sorted, elonogated or shortened, Tuples are often used to store related values that are constant and do not change. A classic example would be genomic co-ordinates.\n","\n","Because they are immutable they are also more memory-efficient and faster to access and process than lists.\n","\n","Tuples are differentiated from lists by using round brackets (parentheses) instead. Because we define a specific format we can access elements like so:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1683732448646,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"32i4KcYAUwb5","outputId":"20dfccd4-01f3-4ffe-bef2-d08c8a7d8e81"},"outputs":[],"source":["# Define a genomic coordinate as a tuple\n","TFBS_coords = ('TFBS01', 1500, 1700)\n","\n","# Print the chromosome name\n","print(TFBS_coords[0])\n","\n","# Print the start position\n","print(TFBS_coords[1])\n","\n","# Print the end position\n","print(TFBS_coords[2])\n"]},{"cell_type":"markdown","metadata":{"id":"TyArylsGQ919"},"source":["Because of the fixed order of tuples there is also a very useful function to quickly assign the values to variables.\n","\n","We know the exact length and order of the tuple elements so we can assign them together on one line:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":275,"status":"ok","timestamp":1683732630933,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"qakt6KEOQ9gu","outputId":"beed1bc3-f462-4a62-d857-b396f026fe8c"},"outputs":[],"source":["# Assign our three value tuple, to three new variable names\n","id, start, end = TFBS_coords\n","\n","# And we can use them immediately\n","print(\"ID:   \", id)\n","print(\"Start:\", start)\n","print(\"End:  \", end)\n","print(\"BED format output:\", id + \":\" + str(start) + \"-\" + str(end))"]},{"cell_type":"markdown","metadata":{"id":"pVHT436sXvYe"},"source":["---\n","\n","# Ranges\n","Ranges are not really a different data type, but lets look at them quickly now as they are a very useful way of creating lists! Really, ranges are a type of automatic list method, where you can define the range and period in your sequence.\n","\n","You can easily create a simple range like so, but note that it is stored as a range-object until you force it to be a list:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":871,"status":"ok","timestamp":1683717654010,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"yEDP7O52Y4jW","outputId":"cca647c4-3c2e-4356-ed54-038e6cb69e20"},"outputs":[],"source":["x = range(20)\n","\n","# As it's own object\n","print(x)\n","\n","# Forced List\n","print(list(x))"]},{"cell_type":"markdown","metadata":{"id":"sTzljupXc-09"},"source":["You can also define the interval that you want your sequence to have"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":186,"status":"ok","timestamp":1683718709545,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"0M6iGIbCc6zW","outputId":"62e32b86-5231-4c40-ba69-6be490b9f1c9"},"outputs":[],"source":["my_timeline = range(0, 20, 3)\n","print(list(my_timeline))"]},{"cell_type":"markdown","metadata":{"id":"73tE45DaZP-w"},"source":["## Exercise - Creating ranges\n","\n","You have taken a flourescence measurement of your culturing algae every **6 minutes for 3 hours** to see the response to a growth stimulation. Lets plot that data on a graph, starting from time 0.\n","\n","Your data is stored in a list variable named ```measurements```\n","\n","1. How many elements are going to be in your list? (You could either calculate it, or count the length of your measurements)\n","2. Create a list object named ```timeline``` using ```range()```. This will be your X-axis\n","3. Run the graph making code to plot your data\n","\n","Note: Remember that list indexes must be inclusive!"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":12,"status":"ok","timestamp":1684697202990,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"ni-XFz3IZrJ4","outputId":"e4e7c436-bce4-46fa-e044-650518e9aa00"},"outputs":[],"source":["# Your data\n","measurements = [0.68, 1.69, 3.37, 5.61, 7.89, 9.86, 10.9, 10.84, 9.86, 8.22, 6.46, 4.69, 3.19, 2.05, 1.24, 0.7, 0.38, 0.2, 0.1, 0.05, 0.02, 0.01, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 1.0, 2.0, 3.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]\n","\n","## Create your timeline\n","timeline =  \n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":449},"executionInfo":{"elapsed":561,"status":"ok","timestamp":1684697209814,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"kdmwllxUUGJ3","outputId":"d91be75f-3288-44a5-cb50-500424d4bc99"},"outputs":[],"source":["# Make a quick graph\n","import matplotlib.pyplot as plt\n","\n","plt.plot(timeline, measurements)\n","plt.xlabel('Time')\n","plt.ylabel('Fluorescence')\n","plt.show()"]},{"cell_type":"markdown","metadata":{"id":"dHQW6joOSzSV"},"source":["---\n","\n","# Two Dimensional (2D) lists \n","\n","2d lists are where each element of the list is a list itself. You can imagine it a little like rows and columns in a spreadsheet or dataframe. \n","\n","Here the first element in each row is a confidence value, the second is the gene ID, and the third is the length of the gene. \n","\n","We can access individual elements of the annotation table using two indices by combining square bracket indexing like with a normal list"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":223,"status":"ok","timestamp":1684697520487,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"y80yavPgVvQm","outputId":"42530ce9-7525-48e8-80bc-8fc4d41a71da"},"outputs":[],"source":["gene_confs = [[0.92, 'MET1', 2205], [0.82, 'EGFR', 1567], [0.93, 'KRAS', 6523], [0.4, 'TP53', 5002], [0.94, 'ROS5', 1999], [0.87, 'BRCA1', 2323]]\n","\n","# Try and find the each of these outputs in the gene_confs list\n","print(gene_confs[1][1])\n","print(gene_confs[3][0])\n","print(gene_confs[5][2])\n","print(gene_confs[5][-1])"]},{"cell_type":"markdown","metadata":{"id":"IN0PIuhlZvKG"},"source":["We can then also add, remove, or sort the list just the same way as if it was a single list. Sort will work on just the first element in each sublist (there are more complex methods to sort on other elements, for later in the course!)\n","\n","One thing to note: There are two sort-related approaches. the ```.sort()``` method changes the list you already have, however ```sorted(list)``` will create a new list. "]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":4,"status":"ok","timestamp":1684697763243,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"tOUm5LI7zEsQ","outputId":"d1783e3d-be3f-43f2-9b4c-299136fbf0c8"},"outputs":[],"source":["new_list = sorted(gene_confs)\n","# Now we have two lists\n","print(gene_confs)\n","print(new_list)\n","\n","# Here we just modify the first list\n","gene_confs.sort(reverse = True)\n","print(gene_confs)"]},{"cell_type":"markdown","metadata":{"id":"tWtcaTq6aOzg"},"source":["#### Advanced: \n","\n","There is also a method named ```zip``` which can combine lists into an array. Incredibly useful!\n","\n","We will come back to this later in the course in more complex examples but works simply like this:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":8,"status":"ok","timestamp":1683734934636,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"epyWZLAGasNA","outputId":"bab36286-da9a-428b-a627-c3531ae78963"},"outputs":[],"source":["bases = [\"A\", \"T\", \"C\", \"G\"]\n","count = [24, 17, 73, 26]\n","\n","combined_base_count = zip(bases, count)\n","\n","print(combined_base_count)\n","print(list(combined_base_count))"]},{"cell_type":"markdown","metadata":{"id":"k4B5OynX93p9"},"source":["Note that it creates tuples, not a list, and also the requirement to force a ```list``` object to make it print. \n","\n","It can also be helpful to do this in reverse, splitting a 2D list into two lists:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":6,"status":"ok","timestamp":1683979100002,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"TWeRorxK9-Jd","outputId":"2692be23-933f-4bda-b36c-0eb86ce97f69"},"outputs":[],"source":["line_exp = [(\"WT\", 98), (\"Mut1\", 76), (\"Mut2\", 86), (\"Mut3\", 79)]\n","\n","line, exp = zip(*line_exp)\n","\n","print(line)\n","print(exp)\n"]},{"cell_type":"markdown","metadata":{"id":"21ou3cQMXQj-"},"source":["## Exercise: Manipulating 2D lists\n","\n","Here we have a list of genes with three values: confidence score, geneID, and gene length\n","\n","1. Sort the gene list by confidence value\n","2. Create a new variable named ```highest``` which contains the gene ID of the list with the highest confidence value and print it\n","3. We have new data! Add the gene ```APC1``` with the conf value ```0.97``` and length ```4287``` and the gene ```COI``` with ```0.72``` and ```1660``` to the list (there are several ways to do this! You could look back at the joining list section)\n","4. Put the three genes with the lowest confidence into a new list named ```low_conf_genes``` and print it.\n","\n","Extension: \n","\n","5. Add a new element into the gene_confs list for the low confidence genes so that they have a 4th element saying ```RETEST```\n","\n","Note: If you have done some programming before you may notice that in several of these examples there are better and quicker ways to achieve the steps by using loops and conditionals. We will cover this in the next session but for now, use the commands we've learned\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":3,"status":"ok","timestamp":1683736018127,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"PF0FvjEqeaZ2","outputId":"0ff6f999-39db-4fc2-cd99-bb7132682a96"},"outputs":[],"source":["gene_confs = [[0.92, 'MET1', 2205], [0.82, 'EGFR', 1567], [0.93, 'KRAS', 6523], [0.4, 'TP53', 5002], [0.94, 'ROS5', 1999], [0.87, 'BRCA1', 2323]]\n","\n","# Your code\n"]},{"cell_type":"markdown","metadata":{"id":"rQ5nBszYVmZT"},"source":["---\n","\n","# Dictionaries\n","\n","A dictionary at first looks like a 2D list but is actually unique because all data is in unordered key-value pairs. They are particularly useful for finding the corresponding pair (like a real dictionary!) when you have one value for example:\n","\n","- ```geneID => sequence```\n","- ```aminoAcid => frequency```\n","- ```site => longitude/lattitude```\n","\n","This example dictionary has a frequency count of each base, where the first element in each pair is the \"key\" reference, and the second element is the \"value\". \n","\n","It's defined in the format of ```'key' : 'value'```, between braces ```{ }``` (curly brackets)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":265,"status":"ok","timestamp":1684698069685,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"rTmf-Cp8WHSi","outputId":"834be94c-b56f-465c-d251-c975f59513de"},"outputs":[],"source":["# Create a dictionary of DNA base counts\n","base_counts = {'A': 101, 'T': 250, 'G': 125, 'C': 92}\n","print(base_counts)"]},{"cell_type":"markdown","metadata":{"id":"cw8OZ_CwWvxT"},"source":["We can then access individual values based on their key just like looking them up in a dictionary, or modify/delete them:"]},{"cell_type":"code","execution_count":90,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":260,"status":"ok","timestamp":1684698071341,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"JMKJ30HtW07q","outputId":"53a85cda-f52f-4645-8121-0ea5a99647ec"},"outputs":[{"name":"stdout","output_type":"stream","text":["Number of Adenine: 101\n","Number of Thymine: 250\n","{'A': 101, 'T': 250, 'G': 125, 'C': 92}\n","{'A': 65, 'T': 250, 'G': 125, 'N': 5}\n","{'A': 65, 'G': 125, 'N': 5}\n"]}],"source":["base_counts = {'A': 101, 'T': 250, 'G': 125, 'C': 92}\n","\n","# Print corresponding value (two methods)\n","print(\"Number of Adenine:\", base_counts.get('A'))\n","print(\"Number of Thymine:\", base_counts['T'])\n","\n","print(base_counts)\n","# Modify the value for a key\n","base_counts['A'] = 65\n","\n","# Add a new key-value pair to the dictionary\n","base_counts['N'] = 5\n","\n","# Remove the 'C' key-value pair from the dictionary\n","base_counts.pop('C')\n","\n","print(base_counts)\n","\n","# There is also the del statement\n","del base_counts['T']\n","print(base_counts)"]},{"cell_type":"markdown","metadata":{},"source":["We can use the square bracket method to return the value of a dictionary key, however the .get() method is usually better because it handles missing data better. Note the difference:"]},{"cell_type":"code","execution_count":92,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Number of Adenine: None\n"]}],"source":["print(\"Number of Adenine:\", base_counts.get('X'))"]},{"cell_type":"code","execution_count":93,"metadata":{},"outputs":[{"ename":"KeyError","evalue":"'X'","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)","\u001b[1;32mc:\\Users\\danie\\Google Drive\\Work\\CompassBioinformatics\\Courses\\IntroToPython\\Notebooks\\Day1\\P34B-Day1-MoreDataTypes.ipynb Cell 50\u001b[0m line \u001b[0;36m1\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Day1/P34B-Day1-MoreDataTypes.ipynb#Y123sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mNumber of Thymine:\u001b[39m\u001b[39m\"\u001b[39m, base_counts[\u001b[39m'\u001b[39;49m\u001b[39mX\u001b[39;49m\u001b[39m'\u001b[39;49m])\n","\u001b[1;31mKeyError\u001b[0m: 'X'"]}],"source":["print(\"Number of Thymine:\", base_counts['X'])"]},{"cell_type":"markdown","metadata":{},"source":["A useful set of methods are ```.keys()``` and ```.values()``` which return dict objects (just like when we made a range). They can be turned into a list using the ```list()``` function in the same way.\n","\n","We can also use the ```.items()``` method to return key:value pairs as a tuple, but that is for more complicated functions we will look at later.\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Print just the values\n","print(base_counts.values())\n","\n","# Print just the keys\n","print(base_counts.keys())"]},{"cell_type":"markdown","metadata":{"id":"KsVBxlMZXqI7"},"source":["---\n","#### Dictionary Example\n","Dictionaries are really important and powerful. Lets look at another example where we can attach specific information to geneIDs, and use the ```.get``` method to search the dictionary.\n","\n","Here we can combine a gene sequence dictionary with our 2D list exercise from earlier"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":225,"status":"ok","timestamp":1684698302821,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"O8UFUTAvXucr","outputId":"e510cc6e-5dbc-4e7a-9369-75d3e76083d6"},"outputs":[],"source":["## The data\n","\n","# Dictionary of gene names and sequences\n","gene_dict = {'BRCA1': 'ATGTTGTCATCGTTGAGCTTTGCTTCCT',\n","             'TP53': 'ATGGAGGAGCCGCAGTCAGATC',\n","             'EGFR': 'ATGACCATCCAAGATGATGGTGTC',\n","             'KRAS': 'ATGACTGAATATAAACTTGTGGTAG',\n","             'BRAF': 'ATGGTCCAGCTTGGACCCACTCC',\n","             'ALK': 'ATGAAGGAGCCCTCAGATTTCTTG',\n","             'RET': 'ATGGGTGGGTTGTCGGAAGATCTT',\n","             'ROS1': 'ATGAGCCACCCAGGTCCCTGTAGT',\n","             'MET': 'ATGGCTTCAAGCTGTTGTCGTGAAGA'}\n","\n","# gene confidence values\n","gene_confs = [[0.92, 'MET1', 2205], [0.82, 'EGFR', 1567], [0.93, 'KRAS', 6523], [0.4, 'TP53', 5002], [0.94, 'ROS5', 1999], [0.87, 'BRCA1', 2323]]\n","\n","# sort and get lowest conf gene ID\n","gene_confs.sort()\n","lowest_gene_conf = gene_confs[0][0]\n","lowest_gene_ID = gene_confs[0][1]\n","\n","print(lowest_gene_ID)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# Search dictionary keys for that ID\n","print(gene_dict.get(lowest_gene_ID))\n","\n","# easier to read!\n","print(\"Gene sequence for\", lowest_gene_ID, \"(Confidence value:\", str(lowest_gene_conf) + \") is:\",  gene_dict.get(lowest_gene_ID))"]},{"cell_type":"markdown","metadata":{},"source":["Lets continue our example from 2D lists and remove the three lowest genes from our dictionary:"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":236,"status":"ok","timestamp":1684698433198,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"jA5MCmWZTGIp","outputId":"12a007be-1ab0-4c20-8468-3f84aed09b13"},"outputs":[],"source":["print(\"gene_dict contains\", len(gene_dict), \"genes\")\n","gene_dict.pop(\"EGFR\")\n","gene_dict.pop(\"TP53\")\n","gene_dict.pop(\"BRCA1\")\n","print(\"gene_dict contains\", len(gene_dict), \"genes\")"]},{"cell_type":"markdown","metadata":{"id":"2-4AuHdaT1Xk"},"source":["Actually, even better idea! Instead of just throwing the data out, lets put the DNA sequences of the lowest three genes from the dictionary into a new list called ```bad_genes``` before they get discarded. \n","\n","Note, what happens if you run this code immediately? Read the error and identify what is wrong."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":214,"status":"ok","timestamp":1684849540464,"user":{"displayName":"","userId":""},"user_tz":-60},"id":"2JQbjR594Ob2","outputId":"c366c781-8780-4aa2-e3ba-438206463d19"},"outputs":[],"source":["bad_genes = []\n","\n","print(\"gene_dict contains\", len(gene_dict), \"genes\")\n","bad_genes.append(gene_dict.pop(\"EGFR\"))\n","bad_genes.append(gene_dict.pop(\"TP53\"))\n","bad_genes.append(gene_dict.pop(\"BRCA1\"))\n","print(\"gene_dict contains\", len(gene_dict), \"genes\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"vfJfdHKn4SCZ"},"outputs":[],"source":["print(bad_genes)\n","print(gene_dict)"]},{"cell_type":"markdown","metadata":{"id":"QKAudoMCYk2t"},"source":["Dictionaries and lists are powerful ways to handle data and often involve going between the two. There are also lots of powerful methods to search throughnot just the keys, but also the values to find relevant keys which can be very useful but first we need to learn a bit about loops and conditionals!"]},{"cell_type":"markdown","metadata":{"id":"xKbU1nRZkgP5"},"source":["## Exercises\n","\n","1. Create a dictionary of bacteria and confluence values using the data:\n","\n","```\n","[\"E. coli\", \"S. aureus\", \"P. aeruginosa\", \"K. pneumoniae\", \"A. baumannii\"]\n","[60, 82, 75, 91, 70]\n","```\n","2. Print the confluence value of *K. pneumoniae* from the dictionary\n","3. A function we haven't used yet is ```sum()```, but it works just like ```len()```. Use both of these to calculate the average confluence of all the samples in the dictionary (what's the easiest way to get all the values?)\n","\n","\n"]},{"cell_type":"code","execution_count":95,"metadata":{"id":"8JOxai9ZlReC"},"outputs":[{"name":"stdout","output_type":"stream","text":["91\n","75.6\n"]}],"source":["# Format reminder\n","bacteria_dict = {\"E. coli\": 60,}"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyORQBISWP15Kmwcqi68fXzD","provenance":[{"file_id":"1cWdsyOsrrZnrXUul6lCw1q88jay7SlsD","timestamp":1683731324320}],"toc_visible":true},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"}},"nbformat":4,"nbformat_minor":0}
