# Practical Python Programming for Biologists
<img align="right" width="328" height="380" src="Images/Headshot-lowres.png">

Course written and delivered by Dr Daniel Pass ([Compass Bioinformatics](https://www.CompassBioinformatics.co.uk)).

These materials are designed for use in live training sessions and therefore have spaces where explanations, demonstrations, and solutions would be discussed and shown. Also there are supplementary presentations in a live course, and not all of the data-files are stored in the github repository. If you're reading these materials outside of a course then please tweet me any feedback [@passdan](https://www.twitter.com/passdan)/[@CompassBioinf](https://www.twitter.com/CompassBioinf) and make me feel popular! If you're interested in joining a live course then please get in touch for upcoming ones.

## Overview 
This is a four-day course designed to provide an introduction to the Python programming language for all biologists and life scientists looking to begin or solidify their coding. The course is intended for people at all levels (students, researchers, postdocs, and even PIs and group leaders!) with little or no prior programming experience. The pace of the course will begin foundational and develop into more complex aspects, with lots of time dedicated to the practice of the concepts covered.

## Course design
There are a lot of coding courses in the world that show (dump) lots of information and concepts on new users where you read and write notes, become very good at copy/paste, and then forget 24 hours after the course finishes. This course is designed with time and space to practice the actual code concepts and become proficient at the fundementals, producing real and useful outputs together. Therefore hopefully making it more likely to be directly relevant for your research. 

It also means that in these materials there are some places where I leave an empty space, or give a project designed to be challenging and requiring 20-30 minutes to solve. It is designed to be hard and make you think and struggle! And then we will work together to solve issues, find solutions, and I will provide "perfect" answers.

This github is read-only and my recomendation is to open [Google Colab](https://colab.research.google.com/) and point it to this repository to run and edit the notebooks yourself. However, any Jupyter environment will work.

# Schedule
## Day 1
- [Introduction, concepts, & data types](Day1/P34B-Day1-Introduction.ipynb)
- [Basic data types (strings and numbers), & Data manipulation](Day1/P34B-Day1-BasicDataTypes.ipynb)
- [More Data Types: Lists (and Tuples & Ranges)](Day1/P34B-Day1-Lists_and_more.ipynb)


[Project: Manipulating DNA sequences - Transcription Factor Binding](Day1/P34B-Day1-project.ipynb)

## Day 2
- [More Data Types: Dictionaries](Day2/P34B-Day1-Dictionaries.ipynb)
- [Conditionals & loops](Day2/P34B-Day2-Loops_and_ifs.ipynb)
- [Advanced string manipulation](Day2/P34B-Day2-AdvancedStrings.ipynb)
 
[Project: Messy data handling](Day2/P34B-Day2-Project-MessyData.ipynb)

## Day 3
- [Organising code with functions](Day3/P34B-Day3-Functions.ipynb)
- [I/O handling](Day3/P34B-Day3-IO.ipynb)
- [Modules and libraries: BioPython](Day3/P34B-Day3-Modules-Biopython.ipynb)

[Project: Chipseq, bespoke file formats, and functional organisation](Day3/P34B-Day3-Project-ChipSeq.ipynb)

## Day 4
- [Modules and libraries: pandas, numPy, & sciPy](Day4/P34B-Day4-Modules-Stats.ipynb)
- [Data visualisation packages (matplotlib/seaborn/plotly)](Day4/P34B-Day4-Data_visualisation.ipynb)
- [Your code as command line programs & argparse](Day4/P34B-Day3-CLI_and_argparse.ipynb)
- [Future Directions](Supplementary/P34B-FutureDirections.md)

[Project: objects and stats - viral genomics](Day3/P34B-Day3-Project-covid_modules.ipynb)

## Supplementary
- [Classes and Object Oriented code](Day3/P34B-Day3-ClassesAndObjects.ipynb)
- [Comprehensions](Supplementary/P34B-Comprehensions.ipynb)
- [Coding in the age of ChatGPT and AI bots](Day4/P34B-Day4-ChatGPT_and_AIs.ipynb)
