{"cells":[{"cell_type":"markdown","metadata":{"id":"O2R8cALMijbx"},"source":["# Practical Python Programming for Biologists\n","Author: Dr. Daniel Pass | www.CompassBioinformatics.com\n","\n","---"]},{"cell_type":"markdown","metadata":{"id":"jFBkgBQlauop"},"source":["# Day 3 Project Challenge - Covid dataset with modules\n","\n","In this exercise we are going to look at a complex dataset (n=126) of covid S protein reference sequences and associated metadata and statistics, and we're going to use some 3rd party modules to perform some higher-level analysis.\n","\n","As we have seen today there are many (many!) ways to approach any dataset. First consider a question that you want to test from this data. If you are more interested in statistical testing then you can focus on the metadata and statistics with pandas/scipy. If you are more interested in sequence bioinformatics then you can work with the fasta data. However you'll likely need both to fully investigate.\n","\n","You have been provided with two files:\n","- ```Day3-Project-nextclade_metadata.csv``` - Data & Metadata of 126 reference covid19 strains and the Spike (S) protein modificiation data\n","- ```Day3-Project-nextclade.fasta``` - Reference fasta sequence file for the corresponding spike proteins\n","\n","Essential steps:\n","1. Read in and filter the metadata file to remove bad and missing data (pandas!)\n","2. Read in the fasta sequence file using biopython. Only include sequences that have passed your step one filter!\n","3. Now perform some analysis!\n","\n","Some example challenges:\n","\n","**Introduction**\n","- A function to output the spike protein sequence and key statistics (GC%, length, etc)\n","- Calculate summary statistics per clade/strain - What is the mean number of substitutions per WHO clade?\n","\n","**Middle**\n","- Calculate the number/size of gap regions (strings of Ns) - does this correspond with the quality assesment?\n","- Test for if binding efficiency correlates with any other variables\n","\n","**Higher**\n","- Output a phylogenetic tree of strains (I think the full dataset may be too computaionally intensive for google Colab, however you could use two/three sequences per clade, or alternatively your own PCs)\n","\n","This project is largely undefined and the intention is to give you a real-life look at approaching a dataset for the first time. Think about your objectives before jumping in! Good luck!\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"BT2qGy1GgRHB"},"outputs":[],"source":[]}],"metadata":{"colab":{"authorship_tag":"ABX9TyPziQhJIWSDHG2p0ZaJaWQX","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
