{"cells":[{"cell_type":"markdown","metadata":{"id":"3K_Sky_6lb3K"},"source":["# Day 3"]},{"cell_type":"markdown","metadata":{},"source":["## Functions"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["True\n","False\n","True\n","Wye True\n","Taf False\n","Aber True\n"]}],"source":["# Write your large_power function here:\n","def big_enough(sites, replicates):\n","  if sites * replicates >= 55:\n","    return True\n","  else:\n","    return False\n","\n","# Testing your function\n","print(big_enough(20, 5))\n","# should print True\n","\n","print(big_enough(5, 6))\n","# should print False\n","\n","print(big_enough(11, 5))\n","# should print True\n","\n","rivers = [[\"Wye\", 20, 5],[\"Taf\", 5, 6],[\"Aber\", 11, 5]]\n","\n","for river in rivers:\n","  print(river[0], big_enough(river[1],river[2]))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["True\n","False\n","True\n"]}],"source":["def big_enough_dict(river):\n","  if rivers.get(river)[0] * rivers.get(river)[1] >= 55:\n","    return True\n","  else:\n","    return False\n","  \n","rivers = {\"Wye\": [20, 5], \"Taf\": [5, 6], \"Aber\": [11, 5]}\n","\n","print(big_enough_dict(\"Wye\"))\n","print(big_enough_dict(\"Taf\"))\n","print(big_enough_dict(\"Aber\"))\n"]},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Forward sequence:\t ATGCGCATGCTAGCTAG\n","Complement sequence:\t TACGCGTACGATCGATC\n"]}],"source":["# Your function\n","def complement(seq):\n","  complement = \"\"\n","  for base in seq:\n","    if base == \"A\":\n","      complement += \"T\"\n","    if base == \"T\":\n","      complement += \"A\"\n","    if base == \"C\":\n","      complement += \"G\"\n","    if base == \"G\":\n","      complement += \"C\"\n","\n","  return complement\n","      \n","\n","# Testing your function\n","for_seq = \"ATGCGCATGCTAGCTAG\"\n","\n","print(\"Forward sequence:\\t\", for_seq)\n","print(\"Complement sequence:\\t\", complement(for_seq))"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Forward sequence:\t ATGCGCATGCTAGCTAG\n","Complement sequence:\t CTAGCTAGCATGCGCAT\n"]}],"source":["base_dict = {\"A\" : \"T\", \"T\" : \"A\", \"C\" : \"G\", \"G\" :\"C\"}\n","\n","def comp_dict(seq):\n","    complement = \"\"\n","    for base in seq:\n","        complement += base_dict.get(base)\n","    return complement\n","\n","# Testing your function\n","for_seq = \"ATGCGCATGCTAGCTAG\"\n","\n","print(\"Forward sequence:\\t\", for_seq)\n","print(\"Complement sequence:\\t\", complement(for_seq))"]},{"cell_type":"markdown","metadata":{"id":"4CepgZYOlkTs"},"source":["## Biopython SeqIO"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":8119,"status":"ok","timestamp":1684793864843,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"mJNSgwIPieUp","outputId":"77f95479-695d-4ff7-8ea7-ab8b3d2e227f"},"outputs":[{"name":"stdout","output_type":"stream","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting BioPython\n","  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m27.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from BioPython) (1.22.4)\n","Installing collected packages: BioPython\n","Successfully installed BioPython-1.81\n"]}],"source":["!pip install BioPython"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["from Bio.Seq import Seq\n","\n","insulin = Seq(gene.replace('\\n',''))\n","insulin_rna = insulin.transcribe()\n","insulin_aa = insulin_rna.translate()\n","\n","print(insulin_rna)\n","for base in insulin_aa:\n","    print(\"-{}-\".format(base), end='')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"IyMZ_Z8ilbVS"},"outputs":[],"source":["from Bio import SeqIO, Seq\n","\n","my_F_list = []\n","my_F_dict = {}\n","\n","for seq_record in SeqIO.parse(\"/content/co1_sequences.fasta\", \"fasta\"):\n","    aa_seq = seq_record.seq.translate()\n","    \n","    if aa_seq.startswith('F'):\n","        my_F_list.append(seq_record.seq)\n","        my_F_dict[seq_record.id] = seq_record.seq\n","\n","\n","for seq in my_F_list:\n","    print(seq)\n","\n","for id, seq in my_F_dict.items():\n","    print(id, seq)\n","\n"]},{"cell_type":"markdown","metadata":{"id":"XBbx3LI9-1ir"},"source":["## Biopython filtering embl/gb"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"PD6gW7Wa-1BL"},"outputs":[],"source":["gene_seq_dict = {}\n","\n","for feature in first_record.features:\n","  if feature.type == \"CDS\":\n","    gene_id = feature.qualifiers['gene'][0]\n","    start = feature.location.start\n","    end = feature.location.end\n","\n","    CDS = first_record.seq[start:end]\n","    gene_seq_dict[gene_id]  = CDS\n","\n","\n","for key, value in gene_seq_dict.items():\n","  print(\">\" + key)\n","  print(value)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jmik7KTyHKM1"},"outputs":[],"source":["from Bio import SeqIO\n","\n","# Read the data in\n","record_iter = SeqIO.parse(\"/content/am181037.embl\", \"embl\")\n","first_record = next(record_iter)\n","\n","for feature in first_record.features:\n","    #print(feature)\n","\n","    if feature.type == \"CDS\":\n","        start = feature.location.start\n","        end = feature.location.end\n","        geneID = feature.qualifiers.get('gene')[0]\n","        seqExtrct = first_record.seq[start:end]\n","\n","        mito_genes[geneID] = seqExtrct\n","        print(geneID, \"\\t\", seqExtrct)"]},{"cell_type":"code","execution_count":7,"metadata":{"id":"0sG-C3DxSd66"},"outputs":[{"ename":"NameError","evalue":"name 'y_abun_WT01' is not defined","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[1;32mc:\\Users\\danie\\Google Drive\\Work\\CompassBioinformatics\\Courses\\IntroToPython\\Notebooks\\Solved\\SolvedExamples-Day3.ipynb Cell 14\u001b[0m line \u001b[0;36m1\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m normalized_arr\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m \u001b[39m# WT averaging\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m y1, y2, y3 \u001b[39m=\u001b[39m y_abun_WT01, y_abun_WT02, y_abun_WT03\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m WT_mean_values \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mmean([normalise_array(y1), normalise_array(y2), normalise_array(y3)], axis\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m \u001b[39m# KO averaging\u001b[39;00m\n","\u001b[1;31mNameError\u001b[0m: name 'y_abun_WT01' is not defined"]}],"source":["def normalise_array(y):\n","    # Calculate the norm of the array\n","    norm = np.linalg.norm(y)\n","\n","    # Normalize the array\n","    normalized_arr = y / norm\n","\n","    return normalized_arr\n","\n","\n","# WT averaging\n","y1, y2, y3 = y_abun_WT01, y_abun_WT02, y_abun_WT03\n","WT_mean_values = np.mean([normalise_array(y1), normalise_array(y2), normalise_array(y3)], axis=0)\n","\n","# KO averaging\n","y1, y2, y3 = y_abun_KO01, y_abun_KO02, y_abun_KO03\n","KO_mean_values = np.mean([normalise_array(y1), normalise_array(y2), normalise_array(y3)], axis=0)\n","\n","plt.plot(x_coords,WT_mean_values)\n","plt.plot(x_coords,KO_mean_values)\n","plt.show()"]},{"cell_type":"markdown","metadata":{},"source":["# pandas/numpy/scipy"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["from scipy.stats import shapiro, mannwhitneyu, kruskal, spearmanr, wilcoxon\n","\n","def test_vs_WTavg(sample):\n","    # Perform Shapiro-Wilk test\n","    stat, p_value = shapiro(all_samples[sample])\n","    print(f\"Normality:\\t\\t   Stat = {round(stat,5)}\\tp = {round(p_value,5)}\")\n","\n","    # Mann-Whitney U Test versus WT mean\n","    stat, p = mannwhitneyu(WT_mean_values, all_samples[sample])\n","    print(f'Mann-Whitney U Test:\\t   Stat = {round(stat,5)}\\tp = {round(p,5)}')\n","\n","    # Spearman Rank Correlation versus WT mean\n","    coef, p = spearmanr(WT_mean_values, all_samples[sample])\n","    print(f'Spearman Rank Correlation: Coeff = {round(coef,5)}\\tp = {round(p,5)}')\n","\n","    # Wilcoxon Signed-Rank Test versus WT mean\n","    stat, p = wilcoxon(WT_mean_values, all_samples[sample])\n","    print(f'Wilcoxon Signed-Rank Test: Stat = {round(stat,5)}\\tp = {round(p,5)}')\n","\n","# Run for all samples\n","for sample in all_samples.keys():\n","    print(sample + \"\\n====\")\n","    test_vs_WTavg(sample)\n","    print()"]},{"cell_type":"markdown","metadata":{"id":"vefSyAeG4LXk"},"source":["## Project - Covid"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":7003,"status":"ok","timestamp":1685010459367,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"NJMjjYwucvJ7","outputId":"60dda761-6e53-4c73-ec06-dd4f85e7549b"},"outputs":[{"name":"stdout","output_type":"stream","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting Biopython\n","  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m35.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from Biopython) (1.22.4)\n","Installing collected packages: Biopython\n","Successfully installed Biopython-1.81\n"]}],"source":["!pip install Biopython"]},{"cell_type":"code","execution_count":11,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":630,"status":"ok","timestamp":1685012509925,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"i1qTXlUCYQWy","outputId":"60317059-d44b-4339-d99c-c48efe3c2a4d"},"outputs":[{"name":"stdout","output_type":"stream","text":["[37.63, 5]\n"]}],"source":["# Read in the fasta file as a biopython object\n","from Bio import SeqIO, Seq\n","import pandas as pd\n","\n","def GC_pc(seq_record):\n","    sequence = seq_record.seq\n","    GC_pc = (sequence.count(\"G\") + sequence.count(\"C\")) / len(sequence) *100\n","    return round(GC_pc,2)\n","\n","def number_of_gaps(seq_record, min_size):\n","    sequence = seq_record.seq\n","    N_sizes = []\n","    counter = 0\n","    for base in sequence:\n","        if base == \"N\":\n","            counter += 1\n","        else:\n","            if counter > min_size:\n","                N_sizes.append(counter)\n","            counter = 0\n","\n","    return len(N_sizes)\n","\n","\n","# Read in the fasta file as a biopython object\n","strain_sequences = []\n","for seq_record in SeqIO.parse(\"C:/Users/danie/Google Drive/Work/CompassBioinformatics/Courses/IntroToPython/ClassData/Day3-Project-nextclade.fasta\", \"fasta\"):\n","  strain_sequences.append(seq_record)\n","\n","\n","findings = {}\n","for record in strain_sequences:\n","    #print(record.id)\n","    findings[record.id] = [GC_pc(record), number_of_gaps(record, 3)]\n","    #print(f\"{record.id} \\t {GC_pc(record)}% \\t {number_of_gaps(record, 3)}\")\n","\n","\n","print(findings.get('ON544567'))\n","\n","\n","\n","import matplotlib.pyplot as plt\n"]},{"cell_type":"code","execution_count":17,"metadata":{},"outputs":[{"ename":"FileNotFoundError","evalue":"[Errno 2] No such file or directory: '/content/Day3-Project-nextclade_metadata.csv'","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)","\u001b[1;32mc:\\Users\\danie\\Google Drive\\Work\\CompassBioinformatics\\Courses\\IntroToPython\\Notebooks\\Solved\\SolvedExamples-Day3.ipynb Cell 22\u001b[0m line \u001b[0;36m4\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X31sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mpandas\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mpd\u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X31sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m# Read in the metadata file as a dataframe\u001b[39;00m\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X31sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mread_csv(\u001b[39m\"\u001b[39;49m\u001b[39m/content/Day3-Project-nextclade_metadata.csv\u001b[39;49m\u001b[39m\"\u001b[39;49m, delimiter\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m;\u001b[39;49m\u001b[39m'\u001b[39;49m)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X31sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m descriptive_stats \u001b[39m=\u001b[39m df\u001b[39m.\u001b[39mdescribe()\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X31sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m descriptive_stats\n","File \u001b[1;32mc:\\Users\\danie\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:948\u001b[0m, in \u001b[0;36mread_csv\u001b[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[0;32m    935\u001b[0m kwds_defaults \u001b[39m=\u001b[39m _refine_defaults_read(\n\u001b[0;32m    936\u001b[0m     dialect,\n\u001b[0;32m    937\u001b[0m     delimiter,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    944\u001b[0m     dtype_backend\u001b[39m=\u001b[39mdtype_backend,\n\u001b[0;32m    945\u001b[0m )\n\u001b[0;32m    946\u001b[0m kwds\u001b[39m.\u001b[39mupdate(kwds_defaults)\n\u001b[1;32m--> 948\u001b[0m \u001b[39mreturn\u001b[39;00m _read(filepath_or_buffer, kwds)\n","File \u001b[1;32mc:\\Users\\danie\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:611\u001b[0m, in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    608\u001b[0m _validate_names(kwds\u001b[39m.\u001b[39mget(\u001b[39m\"\u001b[39m\u001b[39mnames\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m))\n\u001b[0;32m    610\u001b[0m \u001b[39m# Create the parser.\u001b[39;00m\n\u001b[1;32m--> 611\u001b[0m parser \u001b[39m=\u001b[39m TextFileReader(filepath_or_buffer, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwds)\n\u001b[0;32m    613\u001b[0m \u001b[39mif\u001b[39;00m chunksize \u001b[39mor\u001b[39;00m iterator:\n\u001b[0;32m    614\u001b[0m     \u001b[39mreturn\u001b[39;00m parser\n","File \u001b[1;32mc:\\Users\\danie\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1448\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[1;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[0;32m   1445\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39moptions[\u001b[39m\"\u001b[39m\u001b[39mhas_index_names\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m kwds[\u001b[39m\"\u001b[39m\u001b[39mhas_index_names\u001b[39m\u001b[39m\"\u001b[39m]\n\u001b[0;32m   1447\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles: IOHandles \u001b[39m|\u001b[39m \u001b[39mNone\u001b[39;00m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m-> 1448\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_make_engine(f, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mengine)\n","File \u001b[1;32mc:\\Users\\danie\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1705\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[1;34m(self, f, engine)\u001b[0m\n\u001b[0;32m   1703\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m mode:\n\u001b[0;32m   1704\u001b[0m         mode \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m-> 1705\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39m=\u001b[39m get_handle(\n\u001b[0;32m   1706\u001b[0m     f,\n\u001b[0;32m   1707\u001b[0m     mode,\n\u001b[0;32m   1708\u001b[0m     encoding\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[0;32m   1709\u001b[0m     compression\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mcompression\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[0;32m   1710\u001b[0m     memory_map\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mmemory_map\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mFalse\u001b[39;49;00m),\n\u001b[0;32m   1711\u001b[0m     is_text\u001b[39m=\u001b[39;49mis_text,\n\u001b[0;32m   1712\u001b[0m     errors\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mencoding_errors\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mstrict\u001b[39;49m\u001b[39m\"\u001b[39;49m),\n\u001b[0;32m   1713\u001b[0m     storage_options\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49moptions\u001b[39m.\u001b[39;49mget(\u001b[39m\"\u001b[39;49m\u001b[39mstorage_options\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39mNone\u001b[39;49;00m),\n\u001b[0;32m   1714\u001b[0m )\n\u001b[0;32m   1715\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m   1716\u001b[0m f \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mhandles\u001b[39m.\u001b[39mhandle\n","File \u001b[1;32mc:\\Users\\danie\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\pandas\\io\\common.py:863\u001b[0m, in \u001b[0;36mget_handle\u001b[1;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[0;32m    858\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39misinstance\u001b[39m(handle, \u001b[39mstr\u001b[39m):\n\u001b[0;32m    859\u001b[0m     \u001b[39m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[0;32m    860\u001b[0m     \u001b[39m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[0;32m    861\u001b[0m     \u001b[39mif\u001b[39;00m ioargs\u001b[39m.\u001b[39mencoding \u001b[39mand\u001b[39;00m \u001b[39m\"\u001b[39m\u001b[39mb\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m ioargs\u001b[39m.\u001b[39mmode:\n\u001b[0;32m    862\u001b[0m         \u001b[39m# Encoding\u001b[39;00m\n\u001b[1;32m--> 863\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39;49m(\n\u001b[0;32m    864\u001b[0m             handle,\n\u001b[0;32m    865\u001b[0m             ioargs\u001b[39m.\u001b[39;49mmode,\n\u001b[0;32m    866\u001b[0m             encoding\u001b[39m=\u001b[39;49mioargs\u001b[39m.\u001b[39;49mencoding,\n\u001b[0;32m    867\u001b[0m             errors\u001b[39m=\u001b[39;49merrors,\n\u001b[0;32m    868\u001b[0m             newline\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[0;32m    869\u001b[0m         )\n\u001b[0;32m    870\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    871\u001b[0m         \u001b[39m# Binary mode\u001b[39;00m\n\u001b[0;32m    872\u001b[0m         handle \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39m(handle, ioargs\u001b[39m.\u001b[39mmode)\n","\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/content/Day3-Project-nextclade_metadata.csv'"]}],"source":["import pandas as pd\n","\n","# Read in the metadata file as a dataframe\n","df = pd.read_csv(\"/content/Day3-Project-nextclade_metadata.csv\", delimiter=';')\n","\n","descriptive_stats = df.describe()\n","descriptive_stats"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyM7zVNHTmXS7jICFahT4spR","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"}},"nbformat":4,"nbformat_minor":0}
