{"cells":[{"cell_type":"markdown","metadata":{"id":"O2R8cALMijbx"},"source":["# Practical Python Programming for Biologists\n","Author: Dr. Daniel Pass | www.CompassBioinformatics.com\n","\n","---"]},{"cell_type":"markdown","metadata":{"id":"jFBkgBQlauop"},"source":["# Day 3 Project Challenge - Covid dataset with modules\n","\n","In this exercise we are going to look at a complex dataset (n=126) of covid S protein reference sequences and associated metadata and statistics, and we're going to use some 3rd party modules to perform some higher-level analysis.\n","\n","As we have seen today there are many (many!) ways to approach any dataset. First consider a question that you want to test from this data. If you are more interested in statistical testing then you can focus on the metadata and statistics with pandas/scipy. If you are more interested in sequence bioinformatics then you can work with the fasta data. However you'll likely need both to fully investigate.\n","\n","You have been provided with two files:\n","- ```Day3-Project-nextclade_metadata``` - Data & Metadata of 126 reference covid19 strains and the Spike (S) protein modificiation data\n","- ```Day3-Project-nextclade.fasta``` - Reference fasta sequence file for the corresponding spike proteins\n","\n","Essential steps:\n","1. Read in and filter the metadata file to remove bad and missing data (pandas!)\n","2. Read in the fasta sequence file using biopython. Only include sequences that have passed your step one filter!\n","3. Now perform some analysis!\n","\n","Some example challenges:\n","- A function to output the spike protein sequence and key statistics (GC%, length, etc)\n","- Calculate the number/size of gap regions (strings of Ns) - does this correspond with the quality assesment?\n","- A statistical test for if binding efficiency correlates with any other variables\n","- Test variations in sequence substitutions in metadata\n","- Calculate summary statistics per clade/strain\n","- Output a phylogenetic tree of strains (I think the full dataset may be too computaionally intensive for google Colab, however you could use one/two sequences per clade, or alternatively your own PCs)\n","\n","This project is largely undefined and the intention is to give you a real-life look at approaching a dataset for the first time. Think about your objectives before jumping in! Good luck!\n"]},{"cell_type":"markdown","metadata":{},"source":["### Some statistics"]},{"cell_type":"code","execution_count":16,"metadata":{"id":"BT2qGy1GgRHB"},"outputs":[{"name":"stdout","output_type":"stream","text":["Index(['index', 'seqName', 'clade', 'Nextclade_pango', 'partiallyAliased',\n","       'clade_nextstrain', 'clade_who', 'clade_legacy', 'qc.overallScore',\n","       'qc.overallStatus', 'totalSubstitutions', 'totalDeletions',\n","       'totalInsertions', 'totalFrameShifts', 'totalMissing', 'totalNonACGTNs',\n","       'totalAminoacidSubstitutions', 'totalAminoacidDeletions',\n","       'totalAminoacidInsertions', 'totalUnknownAa', 'alignmentScore',\n","       'alignmentStart', 'alignmentEnd', 'coverage', 'substitutions',\n","       'bindingEfficiency'],\n","      dtype='object')\n","   index                                            seqName        clade  \\\n","0      0                                           OX458355          23B   \n","1      1                                      IND/2930/2023          22F   \n","2      2  IMS-10116-CVDP-88333393-399B-4563-A7B1-14DFC5D...          21L   \n","3      3  IMS-10004-CVDP-63A4C3D8-D0DC-4A57-93D5-BB10D58...          22D   \n","4      4                         England/QEUH-3267E648/2022  recombinant   \n","\n","  Nextclade_pango           partiallyAliased clade_nextstrain    clade_who  \\\n","0            FU.2               XBB.1.16.1.2              23B      Omicron   \n","1         XBB.2.3                    XBB.2.3              22F      Omicron   \n","2        CM.8.1.1            BA.2.3.20.8.1.1              21L      Omicron   \n","3          DV.1.1  BA.2.75.3.4.1.1.1.1.1.1.1              22D      Omicron   \n","4           XBC.1                      XBC.1      recombinant  recombinant   \n","\n","    clade_legacy  qc.overallScore qc.overallStatus  ...  \\\n","0  23B (Omicron)        10.406598             good  ...   \n","1  22F (Omicron)         0.010000             good  ...   \n","2  21L (Omicron)       117.495556              bad  ...   \n","3  22D (Omicron)         3.320950             good  ...   \n","4    recombinant         2.457833             good  ...   \n","\n","   totalAminoacidSubstitutions  totalAminoacidDeletions  \\\n","0                         69.0                       13   \n","1                         62.0                        0   \n","2                         70.0                       12   \n","3                         73.0                       12   \n","4                         61.0                       11   \n","\n","   totalAminoacidInsertions  totalUnknownAa  alignmentScore  alignmentStart  \\\n","0                         0           281.0         89122.0               1   \n","1                         0           121.0         88914.0              26   \n","2                         0             0.0         89226.0               1   \n","3                         0           269.0         87874.0             345   \n","4                         1           202.0         89258.0               1   \n","\n","   alignmentEnd  coverage  substitutions  bindingEfficiency  \n","0         29903  0.960840          A405G         194.483571  \n","1         29783  0.984216           C44T         179.308679  \n","2         29903  0.986657          C241T         193.238443  \n","3         29827  0.959904          T670G         201.615149  \n","4         29903  0.979032          C241T         146.829233  \n","\n","[5 rows x 26 columns]\n"]}],"source":["import pandas as pd\n","\n","## Read in metadata\n","df = pd.read_csv(\"C:/Users/danie/Google Drive/Work/CompassBioinformatics/Courses/IntroToPython/ClassData/Day3-Project-nextclade_metadata.csv\")\n","\n","print(df.columns)\n","print(df.head())"]},{"cell_type":"code","execution_count":18,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["clade_who\n","Delta          30.500000\n","Epsilon        14.333333\n","Eta            19.000000\n","Kappa          21.333333\n","Lambda         21.000000\n","Mu             23.666667\n","Omicron        57.452830\n","Theta          20.000000\n","recombinant    45.571429\n","Name: totalAminoacidSubstitutions, dtype: float64\n"]}],"source":["goodSamples = df[df[\"qc.overallStatus\"] == \"good\"]\n","\n","print(goodSamples.groupby(\"clade_who\")[\"totalAminoacidSubstitutions\"].mean())"]},{"cell_type":"code","execution_count":32,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Spearman Rank test totalAminoacidSubstitutions vs binding Efficiency\n","R2= 0.974 (p=0.0)\n","Spearman Rank test totalAminoacidDeletions vs binding Efficiency\n","R2= 0.769 (p=0.0)\n","Spearman Rank test totalAminoacidInsertions vs binding Efficiency\n","R2= 0.032 (p=0.65916)\n","================\n","Spearman Rank test totalAminoacidSubstitutions vs binding Efficiency\n","R2= 0.985 (p=0.0)\n","Spearman Rank test totalAminoacidDeletions vs binding Efficiency\n","R2= 0.735 (p=0.0)\n","Spearman Rank test totalAminoacidInsertions vs binding Efficiency\n","R2= 0.073 (p=0.39732)\n"]}],"source":["from scipy.stats import mannwhitneyu, spearmanr\n","\n","corrs_to_test = ['totalAminoacidSubstitutions', 'totalAminoacidDeletions','totalAminoacidInsertions']\n","\n","# All samples\n","for x in corrs_to_test:\n","    stat, pvalue = spearmanr(df[x], df[\"bindingEfficiency\"])\n","    print(\"Spearman Rank test\", x, \"vs binding Efficiency\")\n","    print(\"R2= {} (p={})\".format(round(stat,3), round(pvalue, 5)))\n","\n","print(\"================\")\n","# 'good' samples\n","for x in corrs_to_test:\n","    stat, pvalue = spearmanr(goodSamples[x], goodSamples[\"bindingEfficiency\"])\n","    print(\"Spearman Rank test\", x, \"vs binding Efficiency\")\n","    print(\"R2= {} (p={})\".format(round(stat,3), round(pvalue, 5)))\n","\n"]},{"cell_type":"markdown","metadata":{},"source":["### loading sequence data"]},{"cell_type":"code","execution_count":33,"metadata":{},"outputs":[],"source":["# Read in the fasta file as a biopython object\n","from Bio import SeqIO, Seq\n","\n","# Read in the fasta file as a biopython object\n","strain_sequences = []\n","for seq_record in SeqIO.parse(\"C:/Users/danie/Google Drive/Work/CompassBioinformatics/Courses/IntroToPython/ClassData/Day3-Project-nextclade.fasta\", \"fasta\"):\n","  strain_sequences.append(seq_record)"]},{"cell_type":"markdown","metadata":{},"source":["#### Functions"]},{"cell_type":"code","execution_count":43,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["GC: 36.26%\tNo. gaps: 6\t['good' 'good' 'good' 'good' 'good']\tOX458355\n","GC: 37.48%\tNo. gaps: 11\t['good' 'good' 'good']\tIND/2930/2023\n","GC: 36.74%\tNo. gaps: 9\t[]\tIMS10004CVDP63A4C3D8D0DC4A5793D5BB10D58E8487\n","GC: 37.79%\tNo. gaps: 7\t['good']\tOP958840\n","GC: 34.26%\tNo. gaps: 12\t['mediocre']\tON629031\n","GC: 34.81%\tNo. gaps: 158\t[]\tUSA/TXCDCASC210279141/2021\n","GC: 11.75%\tNo. gaps: 20\t['bad']\tLR991190\n","GC: 29.97%\tNo. gaps: 90\t[]\tUSA/COCDPHE2007261261/2020\n","GC: 37.53%\tNo. gaps: 19\t['good']\tUSA/TG278209/2020\n","GC: 34.8%\tNo. gaps: 7\t['unknown']\tSYC/G38725/2021\n","GC: 36.71%\tNo. gaps: 7\t['unknown']\tOU221273\n","GC: 37.32%\tNo. gaps: 8\t[]\tUSA/NJPHEL2112068/2021\n","GC: 35.64%\tNo. gaps: 9\t[]\tUSA/WAS5413/2021\n","GC: 37.23%\tNo. gaps: 12\t[]\tUSA/IDCDCASC210011921/2021\n","GC: 36.17%\tNo. gaps: 13\t[]\tUSA/VACDCLC0067096/2021\n","GC: 23.03%\tNo. gaps: 47\t['bad' 'bad']\tBHR/30005866/2021\n","GC: 36.12%\tNo. gaps: 30\t[]\tUSA/CACDPH3000131458/2021\n","GC: 37.14%\tNo. gaps: 7\t[]\tUSA/MACDCBICRSP_6XNVDTWZVHP437HT/2021\n","GC: 37.52%\tNo. gaps: 14\t[]\tUSA/CACDCFG042304/2021\n","GC: 37.12%\tNo. gaps: 6\t['good']\tOU570677\n","GC: 35.36%\tNo. gaps: 16\t[]\tUSA/MACDCBICRSP_M2WXUWBLS72TQFUR/2021\n","GC: 36.99%\tNo. gaps: 9\t[]\tUSA/MACDCBICRSP_S637QBAELKOECBRG/2021\n","GC: 37.77%\tNo. gaps: 6\t[]\tUSA/VACDCASC210374497/2021\n","GC: 37.58%\tNo. gaps: 15\t[]\tUSA/TXCDCASC210376640/2021\n","GC: 36.39%\tNo. gaps: 17\t[]\tUSA/NJCDCASC210001462/2021\n","GC: 35.38%\tNo. gaps: 53\t[]\tUSA/COCDCMMB10752210/2021\n","GC: 36.33%\tNo. gaps: 14\t['good']\tOU819939\n"]}],"source":["def GC_pc(seq_record):\n","    sequence = seq_record.seq\n","    GC_pc = (sequence.count(\"G\") + sequence.count(\"C\")) / len(sequence) *100\n","    return round(GC_pc,2)\n","\n","def number_of_gaps(seq_record, min_size):\n","    sequence = seq_record.seq\n","    N_sizes = []\n","    counter = 0\n","    for base in sequence:\n","        if base == \"N\":\n","            counter += 1\n","        else:\n","            if counter > min_size:\n","                N_sizes.append(counter)\n","            counter = 0\n","\n","    return len(N_sizes)\n","\n","\n","findings = {}\n","for record in strain_sequences:\n","    #print(record.id)\n","    findings[record.id] = [GC_pc(record), number_of_gaps(record, 3)]\n","    #print(f\"{record.id} \\t {GC_pc(record)}% \\t {number_of_gaps(record, 3)}\")\n","\n","\n","for ID in findings.keys():\n","    GC, gaps = findings.get(ID)\n","    seq_qual = df.loc[df['seqName'] == ID, 'qc.overallStatus'].values\n","    if gaps >5:\n","        print(\"GC: {}%\\tNo. gaps: {}\\t{}\\t{}\".format(GC, gaps, seq_qual, ID))"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyPziQhJIWSDHG2p0ZaJaWQX","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"}},"nbformat":4,"nbformat_minor":0}
