{"cells":[{"cell_type":"markdown","metadata":{"id":"3K_Sky_6lb3K"},"source":["# Day 3"]},{"cell_type":"markdown","metadata":{},"source":["## Functions"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["True\n","False\n","True\n","Wye True\n","Taf False\n","Aber True\n"]}],"source":["# Write your large_power function here:\n","def big_enough(sites, replicates):\n","  if sites * replicates >= 55:\n","    return True\n","  else:\n","    return False\n","\n","# Testing your function\n","print(big_enough(20, 5))\n","# should print True\n","\n","print(big_enough(5, 6))\n","# should print False\n","\n","print(big_enough(11, 5))\n","# should print True\n","\n","rivers = [[\"Wye\", 20, 5],[\"Taf\", 5, 6],[\"Aber\", 11, 5]]\n","\n","for river in rivers:\n","  print(river[0], big_enough(river[1],river[2]))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["True\n","False\n","True\n"]}],"source":["def big_enough_dict(river):\n","  if rivers.get(river)[0] * rivers.get(river)[1] >= 55:\n","    return True\n","  else:\n","    return False\n","  \n","rivers = {\"Wye\": [20, 5], \"Taf\": [5, 6], \"Aber\": [11, 5]}\n","\n","print(big_enough_dict(\"Wye\"))\n","print(big_enough_dict(\"Taf\"))\n","print(big_enough_dict(\"Aber\"))\n"]},{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Forward sequence:\t ATGCGCATGCTAGCTAG\n","Complement sequence:\t TACGCGTACGATCGATC\n"]}],"source":["# Your function\n","def complement(seq):\n","  complement = \"\"\n","  for base in seq:\n","    if base == \"A\":\n","      complement += \"T\"\n","    if base == \"T\":\n","      complement += \"A\"\n","    if base == \"C\":\n","      complement += \"G\"\n","    if base == \"G\":\n","      complement += \"C\"\n","\n","  return complement\n","      \n","\n","# Testing your function\n","for_seq = \"ATGCGCATGCTAGCTAG\"\n","\n","print(\"Forward sequence:\\t\", for_seq)\n","print(\"Complement sequence:\\t\", complement(for_seq))"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["Forward sequence:\t ATGCGCATGCTAGCTAG\n","Complement sequence:\t CTAGCTAGCATGCGCAT\n"]}],"source":["base_dict = {\"A\" : \"T\", \"T\" : \"A\", \"C\" : \"G\", \"G\" :\"C\"}\n","\n","def comp_dict(seq):\n","    complement = \"\"\n","    for base in seq:\n","        complement += base_dict.get(base)\n","    return complement\n","\n","# Testing your function\n","for_seq = \"ATGCGCATGCTAGCTAG\"\n","\n","print(\"Forward sequence:\\t\", for_seq)\n","print(\"Complement sequence:\\t\", complement(for_seq))"]},{"cell_type":"markdown","metadata":{},"source":["# IO"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# IO data\n","\n","my_lines = []\n","\n","with open(\"/content/am181037.embl\") as embl:\n","    for line in embl.readlines():\n","        if not line.startswith(' ') and not line.startswith('FT'):\n","            my_lines.append(line)\n","\n","KW_data = []\n","\n","for line in my_lines:\n","    if line.startswith('KW'):\n","        KW_data.append(line)\n","\n","output_file_name = \"VV\" + str(len(KW_data)) + \"_genes.txt\"\n","\n","with open(\"C:/Users/danie/downloads/\" + output_file_name, 'w') as outfile:\n","    print(\"Genes found in am181037\")\n","    for line in KW_data:\n","        outfile.write(line)"]},{"cell_type":"markdown","metadata":{"id":"4CepgZYOlkTs"},"source":["## Biopython SeqIO"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":8119,"status":"ok","timestamp":1684793864843,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"},"user_tz":-60},"id":"mJNSgwIPieUp","outputId":"77f95479-695d-4ff7-8ea7-ab8b3d2e227f"},"outputs":[{"name":"stdout","output_type":"stream","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting BioPython\n","  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m27.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from BioPython) (1.22.4)\n","Installing collected packages: BioPython\n","Successfully installed BioPython-1.81\n"]}],"source":["!pip install BioPython"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["from Bio.Seq import Seq\n","\n","insulin = Seq(gene.replace('\\n',''))\n","insulin_rna = insulin.transcribe()\n","insulin_aa = insulin_rna.translate()\n","\n","print(insulin_rna)\n","for base in insulin_aa:\n","    print(\"-{}-\".format(base), end='')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"IyMZ_Z8ilbVS"},"outputs":[],"source":["from Bio import SeqIO, Seq\n","\n","my_F_list = []\n","my_F_dict = {}\n","\n","for seq_record in SeqIO.parse(\"/content/co1_sequences.fasta\", \"fasta\"):\n","    aa_seq = seq_record.seq.translate()\n","    \n","    if aa_seq.startswith('F'):\n","        my_F_list.append(seq_record.seq)\n","        my_F_dict[seq_record.id] = seq_record.seq\n","\n","\n","for seq in my_F_list:\n","    print(seq)\n","\n","for id, seq in my_F_dict.items():\n","    print(id, seq)\n","\n"]},{"cell_type":"markdown","metadata":{"id":"XBbx3LI9-1ir"},"source":["## Biopython filtering embl/gb"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"PD6gW7Wa-1BL"},"outputs":[],"source":["gene_seq_dict = {}\n","\n","for feature in first_record.features:\n","  if feature.type == \"CDS\":\n","    gene_id = feature.qualifiers['gene'][0]\n","    start = feature.location.start\n","    end = feature.location.end\n","\n","    CDS = first_record.seq[start:end]\n","    gene_seq_dict[gene_id]  = CDS\n","\n","\n","for key, value in gene_seq_dict.items():\n","  print(\">\" + key)\n","  print(value)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jmik7KTyHKM1"},"outputs":[],"source":["from Bio import SeqIO\n","\n","# Read the data in\n","record_iter = SeqIO.parse(\"/content/am181037.embl\", \"embl\")\n","first_record = next(record_iter)\n","\n","for feature in first_record.features:\n","    #print(feature)\n","\n","    if feature.type == \"CDS\":\n","        start = feature.location.start\n","        end = feature.location.end\n","        geneID = feature.qualifiers.get('gene')[0]\n","        seqExtrct = first_record.seq[start:end]\n","\n","        mito_genes[geneID] = seqExtrct\n","        print(geneID, \"\\t\", seqExtrct)"]},{"cell_type":"code","execution_count":7,"metadata":{"id":"0sG-C3DxSd66"},"outputs":[{"ename":"NameError","evalue":"name 'y_abun_WT01' is not defined","output_type":"error","traceback":["\u001b[1;31m---------------------------------------------------------------------------\u001b[0m","\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[1;32mc:\\Users\\danie\\Google Drive\\Work\\CompassBioinformatics\\Courses\\IntroToPython\\Notebooks\\Solved\\SolvedExamples-Day3.ipynb Cell 14\u001b[0m line \u001b[0;36m1\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m normalized_arr\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m \u001b[39m# WT averaging\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m y1, y2, y3 \u001b[39m=\u001b[39m y_abun_WT01, y_abun_WT02, y_abun_WT03\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m WT_mean_values \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mmean([normalise_array(y1), normalise_array(y2), normalise_array(y3)], axis\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/danie/Google%20Drive/Work/CompassBioinformatics/Courses/IntroToPython/Notebooks/Solved/SolvedExamples-Day3.ipynb#X10sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m \u001b[39m# KO averaging\u001b[39;00m\n","\u001b[1;31mNameError\u001b[0m: name 'y_abun_WT01' is not defined"]}],"source":["def normalise_array(y):\n","    # Calculate the norm of the array\n","    norm = np.linalg.norm(y)\n","\n","    # Normalize the array\n","    normalized_arr = y / norm\n","\n","    return normalized_arr\n","\n","\n","# WT averaging\n","y1, y2, y3 = y_abun_WT01, y_abun_WT02, y_abun_WT03\n","WT_mean_values = np.mean([normalise_array(y1), normalise_array(y2), normalise_array(y3)], axis=0)\n","\n","# KO averaging\n","y1, y2, y3 = y_abun_KO01, y_abun_KO02, y_abun_KO03\n","KO_mean_values = np.mean([normalise_array(y1), normalise_array(y2), normalise_array(y3)], axis=0)\n","\n","plt.plot(x_coords,WT_mean_values)\n","plt.plot(x_coords,KO_mean_values)\n","plt.show()"]}],"metadata":{"colab":{"authorship_tag":"ABX9TyM7zVNHTmXS7jICFahT4spR","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.5"}},"nbformat":4,"nbformat_minor":0}
