{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"toc_visible":true,"authorship_tag":"ABX9TyOM1W4zGR9TmSHm1vdDAUoU"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Day 2 - Introduction to Python Programming for Biologists\n","All commands for today's session are found below, except where you see the word ***EXAMPLE*** which may be just non functional psuedo-code to explain a concept.\n","\n","If you see ***Exercises*** then there is a challenge for you to write some code yourself!\n","\n","## Day 2\n","- Conditionals & loops\n","- Handling bioinformatic file formats\n","- Writing functions & object oriented code\n","- Project: Extract, parse and process sequence data from FASTA and GenBank files\n"],"metadata":{"id":"6bLlQkRwJcqY"}},{"cell_type":"markdown","source":["# A quick bit of maths\n","\n","We will always end up doing some calculations in our coding. The classic opperators apply and we have already used (```+ - / *```), but to note some more specific cases:\n","- ```**``` - Exponent (To the power of). \n","\n"," e.g. ```7 ** 2 = 49```\n","- ```%``` - Modulo. Return just the remainder of a division.\n","\n"," e.g. ```17 % 5 = 2```\n","\n","Lets see whether our gene has a perfect amino acid translation. 3 bases per codon so our modulo test should return zero:"],"metadata":{"id":"15PxGLLfOoz_"}},{"cell_type":"code","source":["my_ORF = \"ATGGTACGACTCACATCACATGCATCGTCAACTTC\"\n","\n","print(\"There are\", len(my_ORF) % 3, \"bases not in a codon\") \n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"QCyJJdcFQa0S","executionInfo":{"status":"ok","timestamp":1683799431466,"user_tz":-60,"elapsed":238,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"1998ce5b-2a81-4066-9bde-cc42446210f9"},"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":["There are 2 bases not in a codon\n"]}]},{"cell_type":"markdown","source":["The other methods we will use are modifying a variable with an opperation to it's current value. We do this by combining a maths operator with the equals:\n","  ```+=     -=     *=    /=```\n","\n","For example these perform the same function, but one is quicker and easier:"],"metadata":{"id":"VKlCCllhRWL3"}},{"cell_type":"code","source":["total = 27\n","new_total = total + 3\n","total = new_total\n","\n","print(total)\n","\n","# alternative\n","total += 4\n","print(total)\n","\n","total /= 2       # Note how any division returns a float\n","print(total)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Xv4uBAj-Rv8Y","executionInfo":{"status":"ok","timestamp":1683799894251,"user_tz":-60,"elapsed":350,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"81950b56-197f-4bae-8b21-c8ab2a258249"},"execution_count":10,"outputs":[{"output_type":"stream","name":"stdout","text":["30\n","34\n","17.0\n"]}]},{"cell_type":"markdown","source":["Acting alone they don't do too much, but in the context of loops they are super powerful!"],"metadata":{"id":"VW2SJ5nQSWcf"}},{"cell_type":"markdown","source":["# Looping\n","\n","Really, we know loops are the powerhouse of programming. We have looked at lists, ranges, and dictionaries which have lots of data in them but untill we can run through the full dataset we aren't using them fully.\n","\n","Running through a collection (term used for any kind of itterable set of data) most easily follows a simple pattern:\n","\n","```\n","for item in list:\n","  print(item)\n","```\n","\n","Some other coding languanges you may have seen will require an extra character to define your temporary variable but that's not needed in python. Nice and clean!\n","\n","Python also doesn't use any of those messy parentheses ```[{()}]``` or semi-colons ```;```. You just have to be sure to use tabs/spaces to indent your code blocks.\n","\n","---"],"metadata":{"id":"zxlW4-sbK0zJ"}},{"cell_type":"markdown","source":["Lets use a for loop to perform a calculation on a list of DNA strings. As you can see once you define the temporary variable after the ```for``` then it can be used as you would any variable. \n","\n","**Exercise**: try printing ```seq``` outside of the loop and check what you get back.\n","\n"],"metadata":{"id":"sXwxi86qTtEP"}},{"cell_type":"code","source":["# Calculate the GC content for a list of DNA sequences\n","sequences = [\"ATGCTGACTATATGAATCGTTTGA\", \"GCTAGCTAGCTAGCTAGCTG\", \"CTGACGATCGTACGACTGATCG\",\"ATCGCGCGCGCGCCCTATACG\"]\n","\n","for seq in sequences:\n","    gc_count = seq.count(\"G\") + seq.count(\"C\")\n","    length = len(seq)\n","\n","    print(\"GC%:\", gc_count / length)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"5wo_F7PDwtdP","executionInfo":{"status":"ok","timestamp":1683808154221,"user_tz":-60,"elapsed":503,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"3e0a40b5-6739-4310-f625-0bfeef91de26"},"execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":["GC%: 0.3333333333333333\n","GC%: 0.55\n","GC%: 0.5454545454545454\n","GC%: 0.7142857142857143\n","ATCGCGCGCGCGCCCTATACG\n"]}]},{"cell_type":"markdown","source":["We can also add to global variables which exist outside of the temporary loop. \n","\n","Lets calculate the GC% of all the sequences together. Each time we run through the loop we will add the calculated value to the global variable. Once we've finished with the loop then we can do our calculation on the totals."],"metadata":{"id":"Pt_J7-01yf3-"}},{"cell_type":"code","source":["total_GC_count = 0\n","total_length = 0\n","\n","for seq in sequences:\n","    gc_count = seq.count(\"G\") + seq.count(\"C\")\n","    length = len(seq)\n","\n","    total_GC_count += gc_count\n","    total_length += length\n","\n","    print(\"GC%:\", gc_count / length)\n","\n","print(\"Total GC%:\", total_GC_count / total_length)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"VlPtY3LeyfVO","executionInfo":{"status":"ok","timestamp":1683808441099,"user_tz":-60,"elapsed":228,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"b48ea9d6-6035-4fbf-ea21-4078c60c7036"},"execution_count":18,"outputs":[{"output_type":"stream","name":"stdout","text":["GC%: 0.3333333333333333\n","GC%: 0.55\n","GC%: 0.5454545454545454\n","GC%: 0.7142857142857143\n","Total GC%: 0.5287356321839081\n"]}]},{"cell_type":"markdown","source":["One notable method we can use is looping through a string. It doesn't come in very useful in normal programming, but in bioinformatics we spend a lot of time with strings and sequences!\n","\n","To work on each individual character in a string:"],"metadata":{"id":"3EuF3KpRteXR"}},{"cell_type":"code","source":["sequence = \"ACCCACACCGTGTG\"\n","\n","count = 0\n","\n","for base in sequence:\n","  print(\"Give me\", base, \"!\")\n","  count += 1\n","\n","print(\"Counted bases:\", count)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ZEzvqQ70ttNY","executionInfo":{"status":"ok","timestamp":1683890956051,"user_tz":-60,"elapsed":209,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"bbf3636b-6efd-4fd6-fae9-4811cfc4dcfb"},"execution_count":5,"outputs":[{"output_type":"stream","name":"stdout","text":["Give me A !\n","Give me C !\n","Give me C !\n","Give me C !\n","Give me A !\n","Give me C !\n","Give me A !\n","Give me C !\n","Give me C !\n","Give me G !\n","Give me T !\n","Give me G !\n","Give me T !\n","Give me G !\n","Counted bases: 14\n"]}]},{"cell_type":"markdown","source":["## Exercises - Looping"],"metadata":{"id":"AEv6Uu30eTYI"}},{"cell_type":"code","source":[],"metadata":{"id":"yTB6on0BKRCP"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# While\n","\n","Using the ```while``` function we can create loops in a slightly different way. Basically we say \"if this is true, keep repeating\". For example:"],"metadata":{"id":"mnW4OfmKTq73"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"Qw5BEuqQJODI"},"outputs":[],"source":["count = 0\n","\n","while count < 5:\n","  print(count)       ## Don't forget we start at zero!\n","  count +=1"]},{"cell_type":"markdown","source":["An important thing to watch out for with while loops is accidentally creating infinity! Run this and see what happens. What went wrong?"],"metadata":{"id":"kWfbgIuo03qH"}},{"cell_type":"code","source":["count = 0\n","\n","while count >= 0:\n","  print(count)\n","  count +=1  "],"metadata":{"id":"v6k-dSMB1Fw2"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can also use ```while``` to run through lists. It gives us an advantage in running until a condition is met, or getting and modifying the index value.\n","\n","As we run a while loop for the length of our list, lets use it to access elements of other lists. Note: Here we are assuming our lists are in the same order and the same length! Perhaps not the best assumption to make in the real world! "],"metadata":{"id":"bkC_VRAA2e7f"}},{"cell_type":"code","source":["expression = [2.3, 4.1, 5.7, 3.9, 6.5]\n","genes = ['BRCA1', 'TP53', 'EGFR', 'KRAS', 'MET1']\n","species = ['Arabidopsis', 'Bos taurus', 'C. elegans', 'Drosophila', 'E. coli']\n","threshold = 5.0\n","index = 0\n","\n","while index < len(expression):\n","  print(genes[index], expression[index], species[index])\n","\n","  index += 1"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fih9aYL83WA3","executionInfo":{"status":"ok","timestamp":1683820222353,"user_tz":-60,"elapsed":203,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"b4e20488-8d5e-497e-e204-a9288603b2e2"},"execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":["BRCA1 2.3 Arabidopsis\n","TP53 4.1 Bos taurus\n","EGFR 5.7 C. elegans\n","KRAS 3.9 Drosophila\n","MET1 6.5 E. coli\n"]}]},{"cell_type":"markdown","source":["## Exercises - While "],"metadata":{"id":"f3YKuFcdeW3V"}},{"cell_type":"markdown","source":["# The Big If!\n","\n","Lets take a break from loops for a moment to talk about conditionals. These are basically \"**if** this **then** that\" statements. They are also super powerful and allow us lots of control over our data and code.\n","\n","When working with numerical data if follows normal mathematical symbols:\n","```\n","if x > y:    print(\"x is greater than y\")\n","if x < y:    print(\"x is less than y\")\n","if x == y:   print(\"x is equal to y\")\n","if x != y:   print(\"x is NOT equal to y\")\n","if x >= y:   print(\"x is greater than or equal to y\")\n","if x <= y:   print(\"x is less than or equal to y\")\n","```\n","\n","Note the ```==``` for equals. A common mistake is to use just one ```=``` character but as we know that assigns a variable value instead.\n","\n","Exercise: Test which if statements print when you change the values"],"metadata":{"id":"_by8rrdQ4yC_"}},{"cell_type":"code","source":["x = 5\n","y = 10\n","\n","if x > y:    print(\"x is greater than y\")\n","if x < y:    print(\"x is less than y\")\n","if x == y:   print(\"x is equal to y\")\n","if x != y:   print(\"x is NOT equal to y\")\n","if x >= y:   print(\"x is greater than or equal to y\")\n","if x <= y:   print(\"x is less than or equal to y\")"],"metadata":{"id":"Rraq4b3pogPW"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["A more real example. Lets check if a single value is less than the common 0.05 significance threshold:"],"metadata":{"id":"z6ByRCjkoyFv"}},{"cell_type":"code","source":["gene_data = [[\"TP53\", 0.0001],[\"BRCA1\", 0.002],[\"EGFR\", 0.15],[\"KRAS\", 0.00001],[\"MYC\", 0.05]]\n","\n","# If the value of the first list element, then print the gene ID\n","if gene_data[0][1] < 0.05:\n","  print(gene_data[0][0])\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"AWuYCYIHfazI","executionInfo":{"status":"ok","timestamp":1683820460434,"user_tz":-60,"elapsed":205,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"2b1e16bc-cb3b-4f3e-8ac4-6c60649fa150"},"execution_count":14,"outputs":[{"output_type":"stream","name":"stdout","text":["TP53\n"]}]},{"cell_type":"markdown","source":["Now lets put that in a loop for all our genes"],"metadata":{"id":"syi-ExdPgqhA"}},{"cell_type":"code","source":["for gene in gene_data:\n","  geneID = gene[0]\n","  geneValue = gene[1]\n","\n","  if geneValue < 0.05:\n","    print(geneID])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cSlZKp4LgsXg","executionInfo":{"status":"ok","timestamp":1683820534764,"user_tz":-60,"elapsed":214,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"26c65943-7f5d-409c-b8d7-a3b961115bf6"},"execution_count":17,"outputs":[{"output_type":"stream","name":"stdout","text":["TP53\n","BRCA1\n","KRAS\n"]}]},{"cell_type":"markdown","source":["Exercise: re-write that code to make it only take three lines!\n","\n","---"],"metadata":{"id":"2W6m2n8NhvuH"}},{"cell_type":"code","source":[],"metadata":{"id":"petZyVU4o4v-"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can also use an if statement to search within a string, although there are more efficient ways to do this with larger datasets."],"metadata":{"id":"SpBIjrrAh9ZC"}},{"cell_type":"code","source":["# Check if a DNA sequence contains a specific motif\n","sequence = \"ATGCTGACTGACGTGAGCTAGCTAGCTAGCTAGCTG\"\n","motif = \"GCTAG\"\n","\n","if motif in sequence:\n","    print(\"Motif found in the sequence\")\n"],"metadata":{"id":"qgK1Flrne3Di"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## if/else/elif\n","\n","Ifs are good, but what about if not? Well that's nice and simple using ```else```"],"metadata":{"id":"HCRhRr6wqF8C"}},{"cell_type":"code","source":["# Check if a DNA sequence length is even or odd\n","sequence = \"ATGCTGACTGACGTGAGCTAGCTAGCTAGCTAGCTG\"\n","\n","if len(sequence) % 3 == 0:\n","    print(\"The Open Reading Frame has complete codon usage\")\n","else:\n","    print(\"The Open Reading Frame does not have complete codons\")\n"],"metadata":{"id":"MNpCyr_zqaED"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can then make it more complicated with elif:"],"metadata":{"id":"Mb1ipY2Sqrgs"}},{"cell_type":"code","source":["dna_motifs = [\"ATCG\", \"GCTAGA\", \"ACGAAAT\", \"TAGAC\", \"CGAT\", \"CTGATGGA\", \"GTGGAC\", \"TGCGTTA\", \"AGTC\", \"CGGGAGT\", \"TATGCG\", \"GCATTG\", \"ACCACACACACACACCCC\"]\n","\n","for motif in dna_motifs:\n","  if len(motif) <= 4:\n","      print(\"Short:\\t\", motif)\n","  elif len(motif) <= 6:\n","      print(\"Medium:\\t\", motif)\n","  elif len(motif) <= 8:\n","      print(\"Long:\\t\", motif)\n","  else:\n","      print(\"Woah that's big!\")\n"],"metadata":{"id":"sh3x19ebe6sm"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We can access a dictionary in the same way. An if statement is a useful way to test if the key you are looking for exists. We have a dictionary of SampleIDs and their geographic locations.\n","\n","If we run the code now it'll give an error. \n","Exercise: Put the print function within an IF statement to give an error if the value doesn't exist"],"metadata":{"id":"Gf9iylyQN1Bw"}},{"cell_type":"code","source":["sample_locations = {\n","    'DAP01': (40.7128, -74.0060),\n","    'DAP02': (51.5074, -0.1278),\n","    'DAP03': (48.8566, 2.3522),\n","    'DAP05': (34.0522, -118.2437),\n","    'DAP06': (35.6895, 139.6917)\n","}\n","\n","print(sample_locations[\"DAP04\"])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":208},"id":"zIQ-iyddOCZf","executionInfo":{"status":"error","timestamp":1684184678509,"user_tz":-60,"elapsed":699,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"1917adcc-7dc0-4fd8-cfb0-ee7fc811aca5"},"execution_count":2,"outputs":[{"output_type":"error","ename":"KeyError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)","\u001b[0;32m<ipython-input-2-c7419367a52a>\u001b[0m in \u001b[0;36m<cell line: 9>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m }\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_locations\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"DAP04\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mKeyError\u001b[0m: 'DAP04'"]}]},{"cell_type":"markdown","source":["An alternative method of doing this would be using the ```.get()``` method, which does return a value if missing. Although it doesn't allow you to perform alternative steps like with else/elif"],"metadata":{"id":"NbMcHXWJPbvJ"}},{"cell_type":"code","source":["print(sample_locations.get(\"DAP04\"))"],"metadata":{"id":"0KHZCLVLPxFf","executionInfo":{"status":"ok","timestamp":1684184979194,"user_tz":-60,"elapsed":261,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"399c6596-2486-410c-e508-7781a18c5fa2","colab":{"base_uri":"https://localhost:8080/"}},"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["None\n"]}]},{"cell_type":"markdown","source":["## AND/OR/NOT\n","\n","Just as we get our brains around ifs and elses, we can also combine multiple together. This is useful for situations such as:\n","\n","- If gene binding site contains TATTA and is <100bp from TSS....\n","- If sample count is >100 or in PIs favourite species list\n","- If antibiotic resistance is present but not in my list of already tested....\n","\n","They're good examples, lets see what the code would look like:"],"metadata":{"id":"vhBGH1mMtgUL"}},{"cell_type":"code","source":["# Sample TFBS sequences and their distances from TSS\n","# Note the use of a tuple here for simplicity in the example. \n","gene_binding_sites = [\n","    (\"TFBS1\", \"CGTATTATCG\", 80),\n","    (\"TFBS2\", \"ACGTATCGTATTA\", 120),\n","    (\"TFBS3\", \"CTAGGTTA\", 90),\n","    (\"TFBS4\", \"TATTACGTA\", 40),\n","    (\"TFBS5\", \"ATGCTTACG\", 110)\n","]\n","\n","# Iterate over gene binding sites and check conditions\n","for binding_site in gene_binding_sites:\n","    gene, sequence, distance = binding_site\n","\n","    if \"TATTA\" in sequence and distance < 100:\n","        print(\"Gene\", gene, \"matches the conditions\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pSLhP0p0v55b","executionInfo":{"status":"ok","timestamp":1683824520590,"user_tz":-60,"elapsed":212,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"f5bd3cd6-26cb-4318-f427-1e5d7fea4f56"},"execution_count":52,"outputs":[{"output_type":"stream","name":"stdout","text":["Gene TFBS1 matches the conditions\n","Gene TFBS4 matches the conditions\n"]}]},{"cell_type":"code","source":["# Sample data: animal names and their corresponding sample counts\n","animal_samples = [\n","    (\"Lion\", 7500),\n","    (\"Elephant\", 2100),\n","    (\"Giraffe\", 1250),\n","    (\"Tiger\", 950),\n","    (\"Zebra\", 1800)\n","]\n","\n","# List of PI's favorite species\n","favorite_species = [\"Lion\", \"Giraffe\", \"Cheetah\"]\n","\n","# Iterate over animal samples and check conditions\n","for animal, count in animal_samples:\n","    if count > 2000 or animal in favorite_species:\n","        print(\"Sample\", animal, \"matches the conditions\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cOgNoMELxIVd","executionInfo":{"status":"ok","timestamp":1683824674354,"user_tz":-60,"elapsed":10,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"85867ffe-c4c1-45a3-c4c8-0eb4c391e721"},"execution_count":54,"outputs":[{"output_type":"stream","name":"stdout","text":["Sample Lion matches the conditions\n","Sample Elephant matches the conditions\n","Sample Giraffe matches the conditions\n"]}]},{"cell_type":"code","source":["# antibiotics and resistance presence\n","antibiotics = [\n","    (\"Amoxicillin\", True),\n","    (\"Ciprofloxacin\", False),\n","    (\"Gentamicin\", True),\n","    (\"Tetracycline\", False),\n","    (\"Vancomycin\", True)\n","]\n","\n","already_tested = [\"Amoxicillin\", \"Ciprofloxacin\"]\n","\n","# Test if antibiotic is already in the tested list\n","for antibiotic in antibiotics:\n","  if antibiotic[0] not in already_tested:\n","    print(antibiotic[0])\n","\n","\n","# Test if resistance presence is known, AND NOT in already tested\n","for antibiotic, presence in antibiotics:\n","    if presence and antibiotic not in already_tested:\n","        print(\"Antibiotic\", antibiotic, \"matches the conditions\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"EReDrNNfxhvM","executionInfo":{"status":"ok","timestamp":1683825021132,"user_tz":-60,"elapsed":271,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"3d766422-b902-4b2b-cbcd-bf11511f3b76"},"execution_count":60,"outputs":[{"output_type":"stream","name":"stdout","text":["Gentamicin\n","Tetracycline\n","Vancomycin\n","Antibiotic Gentamicin matches the conditions\n","Antibiotic Vancomycin matches the conditions\n"]}]},{"cell_type":"markdown","source":["## Exercises - ifs, buts, and maybes\n","\n","Lets practice\n","\n","\n","1.   Print the genes with expression greater than 2\n","2.   Print the genes with expression greater than 2 but less than 3\n","3.   Print the genes with expression less than 1 or greater than 3\n","4.   Print the genes with expression less than 2 but not called \"BRACA1\"\n","5.   Print \"Low\", \"Medium\" or \"High\" based on if the value is <2, 2-3, or >3 \n","\n","\n"],"metadata":{"id":"OyjodSs3y2gl"}},{"cell_type":"code","source":["gene_expression = [(\"TP53\", 2.5), (\"BRCA1\", 1.8), (\"EGFR\", 3.2), (\"KRAS\", 0.9), (\"MYC\", 2.7), (\"PTEN\", 1.2), (\"CDKN2A\", 4.5), (\"AKT1\", 2.0), (\"ERBB2\", 1.5), (\"RB1\", 3.8)]\n","\n","for gene, expression in gene_expression:\n","  if ........."],"metadata":{"id":"zPq-SzUdzZAc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Advanced looping"],"metadata":{"id":"ol-FPItjG42y"}},{"cell_type":"markdown","source":["## Loop control\n","If we start running through a loop, it will continue to the end. If our objective is \"Do X for every item in list\" then that is good however, sometimes we don't want that to happen and once we have reached the data we want then we can stop, or skip that turn\n","\n","- ```continue``` - Move on to the next loop immediately - don't do any more code on this itteration\n","- ```break``` - Stop the loop right now, do no more processing"],"metadata":{"id":"dMQeiv3YVTQP"}},{"cell_type":"code","source":["gene_list = [\"RB1\", \"PIK3CA\", \"BRAF\", \"HOX2\", \"JAK2\", \"HOX5\", \"NOTCH1\", \"FLT3\", \"KRAS\", \"EGFR\", \"PTEN\", \"HOX1\", \"BRAF\", \"PIK3CA\", \"HOX9\", \"JAK2\"]\n","\n","print(\"Skipping HOX genes:\")\n","for gene in gene_list:\n","    if \"HOX\" in gene:\n","        continue\n","    print(gene)\n","\n","# Note here it stops before printing anything else on that loop\n","print(\"\\nStopping if I find HOX5:\")\n","for gene in gene_list:\n","    if \"HOX5\" in gene:\n","        break\n","    print(gene)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"3y7vx0vUEQ5c","executionInfo":{"status":"ok","timestamp":1683914137162,"user_tz":-60,"elapsed":235,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}},"outputId":"d7ebe1f4-6596-4f61-eff8-9b795ecb5ed7"},"execution_count":9,"outputs":[{"output_type":"stream","name":"stdout","text":["Skipping HOX genes:\n","RB1\n","PIK3CA\n","BRAF\n","JAK2\n","NOTCH1\n","FLT3\n","KRAS\n","EGFR\n","PTEN\n","BRAF\n","PIK3CA\n","JAK2\n","\n","Stopping if I find HOX5:\n","RB1\n","PIK3CA\n","BRAF\n","HOX2\n","JAK2\n"]}]},{"cell_type":"markdown","source":["# Exercise - Loops, lists, and ifs\n","Take our list of genes and output our list of up and down regulated genes. Create an extra column of whether those genes have been qPCR tested.\n","\n","This exercise is the most dificult we've done and will need you to use lists, loops, ifs, and dictionaries! It is not supposed to be simple and probably will take 10-20 minutes.\n","\n","Suggested steps:\n","\n","1.   Make empty lists for your up/down gene lists\n","2.   Make a loop to run through the list of gene expression values and use if/else to build your lists\n","3.   Itterate through your new lists, search the dictionary, and output in a clean format\n","\n","\n","Extension: If you've completed this and want an extra challenge, filter the data to greater than +/- 2 expression. What % of genes are removed by this filter?\n"],"metadata":{"id":"vfkv7KO9i7jp"}},{"cell_type":"code","source":["# Data for your analysis\n","gene_list = [\"TP53\", \"BRCA1\", \"EGFR\", \"KRAS\", \"MYC\", \"PTEN\", \"CDKN2A\", \"AKT1\", \"ERBB2\", \"RB1\", \"PIK3CA\", \"BRAF\", \"MET\", \"JAK2\", \"MAPK1\", \"NOTCH1\", \"FLT3\", \"KRAS\", \"EGFR\", \"PTEN\", \"TP53\", \"BRAF\", \"PIK3CA\", \"MYC\", \"ERBB2\", \"RB1\", \"AKT1\", \"CDKN2A\", \"MAPK1\", \"JAK2\"]\n","gene_exp = [['TP53', -1.4245], ['BRCA1', 4.0941], ['EGFR', 1.8266], ['KRAS', -0.8996], ['MYC', -2.1188], ['PTEN', -1.8499], ['CDKN2A', 4.2798], ['AKT1', 0.4709], ['ERBB2', 2.144], ['RB1', 4.7541], ['PIK3CA', 2.6291], ['BRAF', 2.1514], ['MET', 0.2558], ['JAK2', 5.0332], ['MAPK1', 5.295], ['NOTCH1', 2.5695], ['FLT3', 5.3211], ['KRAS', 3.064], ['EGFR', 0.1768], ['PTEN', -0.4923], ['TP53', -4.7796], ['BRAF', -3.8257], ['PIK3CA', 0.3357], ['MYC', -0.0917], ['ERBB2', -3.2586], ['RB1', -0.7903], ['AKT1', 0.6729], ['CDKN2A', 5.8131], ['MAPK1', -4.9549], ['JAK2', -0.936]]\n","qPCR_status = {'TP53': False, 'BRCA1': False, 'EGFR': True, 'KRAS': False, 'MYC': True, 'PTEN': True, 'CDKN2A': False, 'AKT1': True, 'ERBB2': False, 'RB1': False, 'PIK3CA': True, 'BRAF': False, 'MET': False, 'JAK2': True, 'MAPK1': False, 'NOTCH1': False, 'FLT3': True}\n"],"metadata":{"id":"rLd2P31djxtx","executionInfo":{"status":"ok","timestamp":1683822503137,"user_tz":-60,"elapsed":308,"user":{"displayName":"Daniel Pass","userId":"05126028130431341848"}}},"execution_count":37,"outputs":[]},{"cell_type":"code","source":[],"metadata":{"id":"fOLCM_TIjTha"},"execution_count":null,"outputs":[]}]}